# Identification of HDV-Like Theta Ribozymes Involved in tRNA-based Recoding of Gut Bacteriophages[![DOI](https://zenodo.org/badge/DOI/10.5281/zenodo.10299930.svg)](https://doi.org/10.5281/zenodo.10299930)


All Code underlaying the discovery of theta ribozymes.

The following packages and external software was used for the analysis:

The following existing/external software was used, though other versions may work: 

genomad v1.3.3, RNArobo v2.1.0, tRNAscan-SE v2.0.9, Python v.3.7.6, Prodigal v2.6.3,  WordCloud v1.5.0, R2R v1.0.6, pandas v1.0.3, NumPy v.1.18.1, Jupyterlab v3.5.0


The whole workflow is split into 3 major sections: 

-Databases for all analysis in annotated databases, motif improvement, taxonomic assignments, coding density and code usage analysis.

-Metagenomic for all analysis on raw reads from over 400000 metagenomic samples.

-Combined: Combining all files to the final tables (S1 and S2), including analysis for unique number of ribozymes, sorting and naming of them.
