{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "run=[\"first\",\"second\",\"third\",\"fourth\"]\n",
    "\n",
    "#run rnarobo for each motif (each iteration and each base: C,T,G,A)\n",
    "bases=[\"C\",\"T\",\"G\",\"A\"]\n",
    "for nr in run:\n",
    "    for bas in bases: \n",
    "        inp=\"/mnt/mnemo6/lukas/anaconda3/bin/rnarobo-2.1.0-linux64 -c --nratio 0.1 /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/\"+nr+\"/\"+nr+\"_\"+bas+\".txt /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/combined.fna >> /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/\"+nr+\"/\"+bas+\"/comb_rnarobo_result_precise.txt\"\n",
    "        subprocess.run(inp, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Identify and cut out all the viral sequences with a ribozyme from the combined FASTA file (first version)\n",
    "\n",
    "\n",
    "# Loop through the bases in the \"bases\" list\n",
    "for bas in bases: \n",
    "    # Create an empty string to store the complete FASTA file\n",
    "    complete_fasta=\"\"\n",
    "    \n",
    "    # Define the file paths for the ribozyme results and the combined FASTA file\n",
    "    filepath=\"/mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/comb_rnarobo_result_precise.txt\"\n",
    "    filepath_fasta=\"/mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/combined.fna\"\n",
    "    \n",
    "    # Open the ribozyme results file\n",
    "    f=open(filepath)\n",
    "    f_l=f.readlines()\n",
    "    \n",
    "    # Initialize variables to store the sequence names and sequences\n",
    "    seqnam=[]\n",
    "    sequence=[]\n",
    "    counter=0\n",
    "    \n",
    "    # Loop through each line of the ribozyme results file\n",
    "    for line in f_l[11:-6]:\n",
    "        # If the line does not start with \"|\", it contains the sequence name\n",
    "        if not line[0]==\"|\":\n",
    "            line=line.strip().split()\n",
    "            seqnam.append(line[2])\n",
    "            counter=1\n",
    "        # If the line starts with \"|\", it contains the sequence itself\n",
    "        elif counter==1:\n",
    "            line=line.strip()\n",
    "            sequence.append(line)\n",
    "            counter=0\n",
    "\n",
    "    # Get the unique sequence names\n",
    "    unique_seqnam=set(seqnam)\n",
    "    \n",
    "    # Loop through the unique sequence names\n",
    "    for seqname in unique_seqnam:\n",
    "        # Define the input command for sed to extract the sequence from the combined FASTA file\n",
    "        inp= \"sed -n '/\"+seqname+\"/,/>/p' \"+filepath_fasta\n",
    "        # Run the input command using the subprocess.run() function\n",
    "        l=subprocess.run(inp,shell=True,capture_output=True,text=True)\n",
    "        # Add the output to the complete FASTA file string\n",
    "        complete_fasta+=l.stdout\n",
    "    \n",
    "    # Split the complete FASTA file string into individual sequences\n",
    "    com_fas=complete_fasta.split(\"\\n\")\n",
    "    fasta_corr=[]\n",
    "    checker=0\n",
    "    fasta_part=\"\"\n",
    "    \n",
    "    # Loop through each line of the complete FASTA file\n",
    "    for line in com_fas[:-1]:\n",
    "        # If the line is empty, skip it\n",
    "        if line==\"\": \n",
    "            continue\n",
    "        # If the line starts with \">\", it contains the sequence name\n",
    "        elif line[0]==\">\":\n",
    "            # If the checker is 0, add the previous sequence to the list of corrected FASTA sequences\n",
    "            if checker==0:\n",
    "                fasta_corr.append(fasta_part)\n",
    "                fasta_part=\"\"\n",
    "                line=line.split()\n",
    "                fasta_corr.append(line[0])\n",
    "                checker=1\n",
    "            # If the checker is 1, set it to 0\n",
    "            else:\n",
    "                checker=0\n",
    "        # If the line does not start with \">\", it contains the sequence itself\n",
    "        else:\n",
    "            fasta_part+=line.strip()\n",
    "\n",
    "    # Add the last sequence to the list of corrected FASTA sequences\n",
    "    fasta_corr.append(fasta_part)\n",
    "    \n",
    "    # Join the corrected FASTA sequences into a single string\n",
    "    fast_seq=\"\\n\".join(fasta_corr)\n",
    "\n",
    "    # Replace any periods in the corrected FASTA sequences with underscores\n",
    "    fast_seq=fast_seq.replace(\".\",\"_\")\n",
    "\n",
    "    # Write the corrected FASTA sequences to a file\n",
    "    f= open(\"/mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_all_viral_sequences_v2.fasta\",\"w+\")\n",
    "    f.write(fast_seq)\n",
    "    f.close()    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Identify and cut out all the viral sequences with a ribozyme from the combined FASTA file (fourth and final version)\n",
    "\n",
    "# Loop through the bases in the \"bases\" list\n",
    "for bas in bases: \n",
    "    # Create an empty string to store the complete FASTA file\n",
    "    complete_fasta=\"\"\n",
    "    \n",
    "    # Define the file paths for the ribozyme results and the combined FASTA file\n",
    "    filepath=\"/mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/fourth/\"+bas+\"/comb_rnarobo_result_precise.txt\"\n",
    "    filepath_fasta=\"/mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/combined.fna\"\n",
    "    \n",
    "    # Open the ribozyme results file\n",
    "    f=open(filepath)\n",
    "    f_l=f.readlines()\n",
    "    \n",
    "    # Initialize variables to store the sequence names and sequences\n",
    "    seqnam=[]\n",
    "    sequence=[]\n",
    "    counter=0\n",
    "    \n",
    "    # Loop through each line of the ribozyme results file\n",
    "    for line in f_l[11:-6]:\n",
    "        # If the line does not start with \"|\", it contains the sequence name\n",
    "        if not line[0]==\"|\":\n",
    "            line=line.strip().split()\n",
    "            seqnam.append(line[2])\n",
    "            counter=1\n",
    "        # If the line starts with \"|\", it contains the sequence itself\n",
    "        elif counter==1:\n",
    "            line=line.strip()\n",
    "            sequence.append(line)\n",
    "            counter=0\n",
    "\n",
    "    # Get the unique sequence names\n",
    "    unique_seqnam=set(seqnam)\n",
    "    \n",
    "    # Loop through the unique sequence names\n",
    "    for seqname in unique_seqnam:\n",
    "        # Define the input command for sed to extract the sequence from the combined FASTA file\n",
    "        inp= \"sed -n '/\"+seqname+\"/,/>/p' \"+filepath_fasta\n",
    "        # Run the input command using the subprocess.run() function\n",
    "        l=subprocess.run(inp,shell=True,capture_output=True,text=True)\n",
    "        # Add the output to the complete FASTA file string\n",
    "        complete_fasta+=l.stdout\n",
    "    \n",
    "    # Split the complete FASTA file string into individual sequences\n",
    "    com_fas=complete_fasta.split(\"\\n\")\n",
    "    fasta_corr=[]\n",
    "    checker=0\n",
    "    fasta_part=\"\"\n",
    "    \n",
    "    # Loop through each line of the complete FASTA file\n",
    "    for line in com_fas[:-1]:\n",
    "        # If the line is empty, skip it\n",
    "        if line==\"\": \n",
    "            continue\n",
    "        # If the line starts with \">\", it contains the sequence name\n",
    "        elif line[0]==\">\":\n",
    "            # If the checker is 0, add the previous sequence to the list of corrected FASTA sequences\n",
    "            if checker==0:\n",
    "                fasta_corr.append(fasta_part)\n",
    "                fasta_part=\"\"\n",
    "                line=line.split()\n",
    "                fasta_corr.append(line[0])\n",
    "                checker=1\n",
    "            # If the checker is 1, set it to 0\n",
    "            else:\n",
    "                checker=0\n",
    "        # If the line does not start with \">\", it contains the sequence itself\n",
    "        else:\n",
    "            fasta_part+=line.strip()\n",
    "\n",
    "    # Add the last sequence to the list of corrected FASTA sequences\n",
    "    fasta_corr.append(fasta_part)\n",
    "    \n",
    "    # Join the corrected FASTA sequences into a single string\n",
    "    fast_seq=\"\\n\".join(fasta_corr)\n",
    "\n",
    "    # Replace any periods in the corrected FASTA sequences with underscores\n",
    "    fast_seq=fast_seq.replace(\".\",\"_\")\n",
    "\n",
    "    # Write the corrected FASTA sequences to a file\n",
    "    f= open(\"/mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/fourth/\"+bas+\"/combined_all_viral_sequences_v2.fasta\",\"w+\")\n",
    "    f.write(fast_seq)\n",
    "    f.close()    \n",
    "    \n"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
