{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Import the subprocess, pandas, and numpy libraries\n",
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Create a list of the four DNA bases\n",
    "bases=[\"C\",\"T\",\"G\",\"A\"]\n",
    "\n",
    "#Perform tRNA search for the first (delta ribozymes) false positive dataset\n",
    "# Loop through the bases in the \"bases\" list\n",
    "for bas in bases: \n",
    "    # Define the input command for tRNAscan-SE with the current base and file paths\n",
    "    inp2=\"/mnt/mnemo6/lukas/anaconda3/envs/sar3/bin/tRNAscan-SE -G /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_all_viral_sequences_v2.fasta -o /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_trna_scan_v2.txt -f /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_trna_scan_secondary_struc.txt\"\n",
    "    \n",
    "    # Run the input command using the subprocess.run() function: Obtain the tRNAs in all phage contigs associated with drzs\n",
    "    subprocess.run(inp2, shell=True)\n",
    "\n",
    "\n",
    "#Perform tRNA search for the fourth (final motif, theta ribozymes) false positive dataset\n",
    "# Loop through the bases in the \"bases\" list\n",
    "for bas in bases: \n",
    "    # Define the input command for tRNAscan-SE with the current base and file paths\n",
    "    inp2=\"/mnt/mnemo6/lukas/anaconda3/envs/sar3/bin/tRNAscan-SE -G /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_all_viral_sequences_v2.fasta -o /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_trna_scan_v2.txt -f /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_trna_scan_secondary_struc.txt\"\n",
    "    \n",
    "    # Run the input command using the subprocess.run() function: Obtain the tRNAs in all phage contigs associated with drzs\n",
    "    subprocess.run(inp2, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "first C 840 105\n",
      "first T 424 49\n",
      "first G 301 2\n",
      "first A 284 12\n",
      "second C 508 97\n",
      "second T 160 38\n",
      "second G 48 2\n",
      "second A 64 2\n",
      "third C 193 76\n",
      "third T 42 30\n",
      "third G 3 1\n",
      "third A 5 1\n",
      "fourth C 302 127\n",
      "fourth T 53 34\n",
      "fourth G 8 5\n",
      "fourth A 11 3\n"
     ]
    }
   ],
   "source": [
    "# Define a list of the four DNA bases\n",
    "bases=[\"C\",\"T\",\"G\",\"A\"]\n",
    "\n",
    "# Define a list of the four sequencing runs\n",
    "run=[\"first\",\"second\",\"third\",\"fourth\"]\n",
    "\n",
    "# Create an empty dictionary to store the results\n",
    "wholedic={}\n",
    "\n",
    "# Create sub-dictionaries for tRNA and ribozyme results, as well as a dictionary for the unique sequences\n",
    "wholedic[\"trna\"]={}\n",
    "wholedic[\"ribo\"]={}\n",
    "wholedic[\"dfs\"]={}\n",
    "\n",
    "# Loop through the sequencing runs in the \"run\" list\n",
    "for nr in run:\n",
    "    # Create sub-dictionaries for tRNA and ribozyme results, as well as a dictionary for the unique sequences, for the current sequencing run\n",
    "    wholedic[\"trna\"][nr]={}\n",
    "    wholedic[\"ribo\"][nr]={}\n",
    "    wholedic[\"dfs\"][nr]={}\n",
    "    \n",
    "    # Loop through the bases in the \"bases\" list\n",
    "    for bas in bases: \n",
    "        # Open the ribozyme results file for the current base and sequencing run\n",
    "        rnarobo_=open(\"/mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/\"+nr+\"/\"+bas+\"/comb_rnarobo_result_precise.txt\")\n",
    "        rnarobo=rnarobo_.readlines()\n",
    "        rnarobo_.close()\n",
    "        \n",
    "        # If the current sequencing run is the fourth one, open the tRNA results file for the current base and sequencing run\n",
    "        if nr==\"fourth\":\n",
    "            trna_=open(\"/mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/fourth/\"+bas+\"/combined_trna_scan_v2.txt\")\n",
    "            trna=trna_.readlines()\n",
    "            trna_.close()\n",
    "        # If the current sequencing run is not the fourth one, open the tRNA results file for the first sequencing run and the current base\n",
    "        else:\n",
    "            trna_=open(\"/mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_trna_scan_v2.txt\")\n",
    "            trna=trna_.readlines()\n",
    "            trna_.close()\n",
    "\n",
    "        # Create a dictionary with tRNA sequence names as keys and starting positions as values\n",
    "        dicti={}\n",
    "        for line in trna[3:]:\n",
    "            line=line.strip().split()\n",
    "            if line[0] in dicti:\n",
    "                dicti[line[0]].append(int(line[3]))\n",
    "            else:\n",
    "                dicti[line[0]]=[int(line[3])]\n",
    "\n",
    "        # Create a dictionary with ribozyme sequence names as keys and starting positions as values, and a list of unique ribozyme sequences\n",
    "        dicti_r={}\n",
    "        ribis=[]\n",
    "        for line in rnarobo[11:-6]:\n",
    "            if line[0]!=\"|\":\n",
    "                line=line.replace(\".\",\"_\")\n",
    "                line=line.strip().split()\n",
    "                if line[2] in dicti_r:\n",
    "                    dicti_r[line[2]].append(int(line[0]))\n",
    "                else:\n",
    "                    dicti_r[line[2]]=[int(line[0])]\n",
    "            # If the current line is a unique ribozyme sequence, append it to the list of unique ribozyme sequences\n",
    "            if line[0]==\"|\":\n",
    "                lin_stripped=\"\".join(line.strip().split(\"|\"))\n",
    "                if lin_stripped not in ribis:\n",
    "                    ribis.append(lin_stripped)\n",
    "        \n",
    "        # Store the tRNA and ribozyme dictionaries for the current base and sequencing run in the \"wholedic\" dictionary\n",
    "        wholedic[\"trna\"][nr][bas]=dicti\n",
    "        wholedic[\"ribo\"][nr][bas]=dicti_r\n",
    "        \n",
    "        # Find the ribozymes that match with tRNAs within a certain range of positions\n",
    "        matching=[]\n",
    "        for key in dicti_r:\n",
    "            if key in dicti:\n",
    "                for value in dicti_r[key]:\n",
    "                    for value2 in dicti[key]:\n",
    "                        if value-value2 in range(-5,6):\n",
    "                            matching.append([key,value,value2])\n",
    "        \n",
    "        # Remove duplicate matches\n",
    "        newm=[]\n",
    "        for mat in matching:\n",
    "            if mat not in newm:\n",
    "                newm.append(mat)\n",
    "        \n",
    "        # Create a DataFrame to store the results\n",
    "        lets_do_this=pd.DataFrame(columns=[\"ID\",\"trna_nr\",\"trna_start\",\"trna_end\",\"trna_type\",\"codon\",\"ins_start\",\"ins_end\",\"trna_confidence\",\"psuedo\",\"rnarobo_start\",\"rnarobo_end\",\"rnarobo_seq\"])\n",
    "        check=0\n",
    "        duplicates=[]\n",
    "        for numbr in range(len(newm)):\n",
    "            match=newm[numbr]\n",
    "            lin_df=\"\"\n",
    "            for line in trna[3:]:\n",
    "                line=line.strip().split()\n",
    "                if match[0]==line[0]:\n",
    "                    if len(lin_df.split())==0:\n",
    "                        if str(match[2])==line[3]:\n",
    "                            lin_df+=\" \".join(line)+\" \"\n",
    "                            if lin_df[-2]!=\"o\":\n",
    "                                lin_df+=\"NA \"\n",
    "                    else: \n",
    "                        duplicates.append(line)\n",
    "            for line in rnarobo[11:-6]:\n",
    "                line=line.replace(\".\",\"_\")\n",
    "                line=line.strip().split()\n",
    "                if len(lin_df.split())<13:\n",
    "                    if check==1:\n",
    "                        lin_df+=str(line[0])\n",
    "                        check=0\n",
    "                    if str(match[1])==line[0]:\n",
    "                        lin_df+=\" \".join(line[:2])\n",
    "                        lin_df+=\" \"\n",
    "                        check=1\n",
    "            lin_f=lin_df.split()\n",
    "            lets_do_this.loc[numbr]=np.array(lin_f)\n",
    "        \n",
    "        # Store the unique ribozyme sequences for the current base and sequencing run in the \"wholedic\" dictionary\n",
    "        wholedic[\"dfs\"][nr][bas]=ribis\n",
    "        \n",
    "        # Print the number of unique delta ribozymes and unique trna associated ribozyme sequences found for the current base and sequencing run\n",
    "        ln=0\n",
    "        for x in dicti_r:\n",
    "            ln+=len(dicti_r[x])\n",
    "        print(nr, bas,len(ribis),len(lets_do_this[\"rnarobo_seq\"].unique()))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#Check which of the unique sequences are foun in both the true als well als the false positive search. If present in both, count them as true positives\n",
    "# Define a list of the four sequencing runs\n",
    "run=[\"first\",\"second\",\"third\",\"fourth\"]\n",
    "\n",
    "# Define a list of the three DNA bases other than \"C\"\n",
    "bases=[\"T\",\"G\",\"A\"]\n",
    "\n",
    "#check for the duplicate hits (e.g. false positives which where found in both CNRA* as well as TNRA*)\n",
    "for nr in run:\n",
    "    for bas in bases[:]:\n",
    "        dictiT_r=wholedic[\"dfs\"][nr][bas]\n",
    "        dicti_r=wholedic[\"dfs\"][nr][\"C\"]\n",
    "        newm=[]\n",
    "        #now lets see which of the unique ribozymes of the dataframe are also found in the C one\n",
    "        for y in set(dictiT_r):\n",
    "            if y in set(dicti_r):\n",
    "                newm.append(y)\n",
    "        print(nr,bas,len(newm))\n",
    "\n",
    "\n",
    "\n",
    "#to get the final number of unique sequences identified from each motif, we need to subtract the number of \"duplicates\" from the number of unique ribozymes obtained prior"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
