{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import numpy as np\n",
    "import pandas as pd\n",
    "from collections import Counter\n",
    "import subprocess\n",
    "\n",
    "# Open the file containing the tRNA scan results\n",
    "trna_=open(\"combined_trna_scan_v2.txt\")\n",
    "trna=trna_.readlines()\n",
    "trna_.close()\n",
    "\n",
    "# Open the file containing the ribozyme scan results\n",
    "rnarobo_=open(\"rnarobo_result_dereplicated.txt\")\n",
    "rnarobo=rnarobo_.readlines()\n",
    "rnarobo_.close()\n",
    "\n",
    "# Create a dictionary to store the starting positions of each tRNA\n",
    "dicti={}\n",
    "# Loop through each line of the tRNA scan results, skipping the header\n",
    "for line in trna[3:]:\n",
    "    # Split the line into a list of values\n",
    "    line=line.strip().split()\n",
    "    # If the sequence name is already in the dictionary, append the starting position to the existing list\n",
    "    if line[0] in dicti:\n",
    "        dicti[line[0]].append(int(line[3]))\n",
    "    # If the sequence name is not in the dictionary, create a new list with the starting position\n",
    "    else:\n",
    "        dicti[line[0]]=[int(line[3])]\n",
    "\n",
    "# Create a dictionary to store the starting positions of each ribozyme\n",
    "dicti_r={}\n",
    "# Loop through each line of the ribozyme scan results, skipping the header and footer\n",
    "for line in rnarobo[11:-6]:\n",
    "    # If the line does not start with \"|\", it contains ribozyme information\n",
    "    if line[0]!=\"|\":\n",
    "        # Replace any periods in the line with underscores\n",
    "        line=line.replace(\".\",\"_\")\n",
    "        # Split the line into a list of values\n",
    "        line=line.strip().split()\n",
    "        # If the scaffold ID is already in the dictionary, append the starting position to the existing list\n",
    "        if line[2] in dicti_r:\n",
    "            dicti_r[line[2]].append(int(line[0]))\n",
    "        # If the scaffold ID is not in the dictionary, create a new list with the starting position\n",
    "        else:\n",
    "            dicti_r[line[2]]=[int(line[0])]\n",
    "\n",
    "# Create an empty list to store the matching tRNA and ribozyme pairs\n",
    "matching=[]\n",
    "# Loop through each scaffold ID in the ribozyme dictionary\n",
    "for key in dicti_r:\n",
    "    # If the scaffold ID is also in the tRNA dictionary\n",
    "    if key in dicti:\n",
    "        # Loop through each ribozyme starting position for the current scaffold ID\n",
    "        for value in dicti_r[key]:\n",
    "            # Loop through each tRNA starting position for the current sequence name\n",
    "            for value2 in dicti[key]:\n",
    "                # If the difference between the ribozyme and tRNA starting positions is within 5 nucleotides\n",
    "                if value-value2 in range(-5,6):\n",
    "                    # Append the scaffold ID, ribozyme starting position, and tRNA starting position to the matching list\n",
    "                    matching.append([key,value,value2])\n",
    "\n",
    "# Display the matching list\n",
    "matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Create an empty DataFrame with the specified column names\n",
    "lets_do_this=pd.DataFrame(columns=[\"ID\",\"trna_nr\",\"trna_start\",\"trna_end\",\"trna_type\",\"codon\",\"ins_start\",\"ins_end\",\"trna_confidence\",\"psuedo\",\"rnarobo_start\",\"rnarobo_end\",\"rnarobo_seq\"])\n",
    "\n",
    "# Initialize a counter and an empty list for duplicate entries\n",
    "check=0\n",
    "duplicates=[]\n",
    "\n",
    "# Loop through each matching ribozyme and tRNA pair\n",
    "for nr in range(len(newm)):\n",
    "    match=newm[nr]\n",
    "    lin_df=\"\"\n",
    "    \n",
    "    # Loop through each line of the tRNA scan results, skipping the header\n",
    "    for line in trna[3:]:\n",
    "        # Split the line into a list of values\n",
    "        line=line.strip().split()\n",
    "        \n",
    "        # If the sequence name matches the current ribozyme and the starting position matches the current tRNA\n",
    "        if match[0]==line[0]:\n",
    "            if len(lin_df.split())==0:\n",
    "                if str(match[2])==line[3]:\n",
    "                    lin_df+=\" \".join(line)+\" \"\n",
    "                    if lin_df[-2]!=\"o\":\n",
    "                        lin_df+=\"NA \"\n",
    "            else: \n",
    "                duplicates.append(line)\n",
    "    \n",
    "    # Loop through each line of the ribozyme scan results, skipping the header and footer\n",
    "    for line in rnarobo[11:-6]:\n",
    "        # Replace any periods in the line with underscores\n",
    "        line=line.replace(\".\",\"_\")\n",
    "        # Split the line into a list of values\n",
    "        line=line.strip().split()\n",
    "        \n",
    "        # If the current line contains ribozyme information\n",
    "        if len(line)>1:\n",
    "            # If the scaffold ID matches the current ribozyme and the starting position matches the current ribozyme\n",
    "            if match[0]==line[2]:\n",
    "                if str(match[1])==line[0]:\n",
    "                    lin_df+=\" \".join(line[:2])\n",
    "                    lin_df+=\" \"\n",
    "                    check=1\n",
    "    \n",
    "    # Split the final string into a list of values and add it as a new row to the DataFrame\n",
    "    lin_f=lin_df.split()\n",
    "    lets_do_this.loc[nr]=np.array(lin_f)\n",
    "\n",
    "# Save the intermediate results to a file and reload it as a pandas DataFrame\n",
    "df1str=lets_do_this.to_csv(sep=\"\\t\")\n",
    "f= open(\"comb_trna_v2_plus_ribo.csv\",\"w+\")\n",
    "f.write(df1str)\n",
    "f.close()\n",
    "fff=pd.read_csv(\"comb_trna_v2_plus_ribo.csv\",sep=\"\\t\",index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import pandas as pd\n",
    "\n",
    "# Add two new columns to the DataFrame and initialize them with empty strings\n",
    "fff[\"trnaseq\"]= \" \"\n",
    "fff[\"ribozyme_control\"]= \" \"\n",
    "\n",
    "# Loop through each row in the DataFrame\n",
    "for ID in range(len(fff)):\n",
    "    # Select the correct contig/genome from the rnarobo hit\n",
    "    inp=\"grep -A1 -m1 '\"+fff.iloc[ID,0]+\"' combined_all_viral_sequences_v2.fasta\" \n",
    "    res=subprocess.run(inp,shell=True, capture_output=True, text=True)\n",
    "    sequ=res.stdout.split()[-1]\n",
    "    start=fff.iloc[ID,2]-1\n",
    "    end=fff.iloc[ID,3]\n",
    "    start2=fff.iloc[ID,10]-1\n",
    "    end2=fff.iloc[ID,11]\n",
    "    \n",
    "    # If the start position is less than the end position\n",
    "    if start < end:\n",
    "        # Extract the tRNA and ribozyme sequences\n",
    "        trna=sequ[start:end]\n",
    "        rnarobo=sequ[start2:end2]\n",
    "        fff.iloc[ID,-2]=trna\n",
    "        fff.iloc[ID,-1]=rnarobo\n",
    "    # If the end position is less than the start position\n",
    "    if end < start:\n",
    "        # Reverse the sequence and extract the tRNA and ribozyme sequences\n",
    "        end-=1\n",
    "        start+=1\n",
    "        end2-=1\n",
    "        start2+=1\n",
    "        trna_rev=sequ[end:start]\n",
    "        ribo_rev=sequ[end2:start2]\n",
    "        trna=[]\n",
    "        rnarobo=[]\n",
    "        # Transcribe the reversed tRNA and ribozyme sequences back into the correct format\n",
    "        for nuc in range(len(trna_rev)-1,-1,-1):\n",
    "            if trna_rev[nuc]==\"A\":\n",
    "                trna.append(\"T\")\n",
    "            if trna_rev[nuc]==\"T\":\n",
    "                trna.append(\"A\")\n",
    "            if trna_rev[nuc]==\"G\":\n",
    "                trna.append(\"C\")\n",
    "            if trna_rev[nuc]==\"C\":\n",
    "                trna.append(\"G\")\n",
    "            if trna_rev[nuc]==\"N\":\n",
    "                trna.append(\"N\")\n",
    "        for nuc in range(len(ribo_rev)-1,-1,-1):\n",
    "            if ribo_rev[nuc]==\"A\":\n",
    "                rnarobo.append(\"T\")\n",
    "            if ribo_rev[nuc]==\"T\":\n",
    "                rnarobo.append(\"A\")\n",
    "            if ribo_rev[nuc]==\"G\":\n",
    "                rnarobo.append(\"C\")\n",
    "            if ribo_rev[nuc]==\"C\":\n",
    "                rnarobo.append(\"G\")\n",
    "            if ribo_rev[nuc]==\"N\":\n",
    "                rnarobo.append(\"N\")\n",
    "        trna=\"\".join(trna)\n",
    "        rnarobo=\"\".join(rnarobo)\n",
    "        fff.iloc[ID,-2]=trna\n",
    "        fff.iloc[ID,-1]=rnarobo\n",
    "\n",
    "# Check if the control and actual sequences are the same\n",
    "for x in range(len(fff)):\n",
    "    seq1=fff.loc[x,\"rnarobo_seq\"]\n",
    "    seq3=\"\".join(seq1.split(\"|\"))\n",
    "    seq2=fff.loc[x,\"ribozyme_control\"]\n",
    "    if seq2!=seq3:\n",
    "        print(\"something wrong\",x)\n",
    "        print(seq2)\n",
    "        print(seq3)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "f=fff.copy(deep=True)\n",
    "#create new columns for study origin\n",
    "f.loc[:,\"study_origin\"]=\"\"\n",
    "\n",
    "#manually add the study origin\n",
    "\n",
    "f.loc[0:59,\"study_origin\"]=\"1\"\n",
    "f.loc[59:299,\"study_origin\"]=\"2\"\n",
    "#three: not found\n",
    "f.loc[299:315,\"study_origin\"]=\"4\"\n",
    "f.loc[315:679,\"study_origin\"]=\"5\"\n",
    "f.loc[679:680,\"study_origin\"]=\"6\"\n",
    "#7 not present\n",
    "f.loc[681:,\"study_origin\"]=\"8\"\n",
    "f[\"study_origin\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Unique ribos: 126\n"
     ]
    }
   ],
   "source": [
    "print (\"Unique ribos: \"+str(len(f[\"ribozyme_control\"].unique())))\n",
    "print (\"Unique trnas: \"+str(len(f[\"trnaseq\"].unique())))\n",
    "#combinme ribo+trnaseq\n",
    "f[\"comb_seq\"]=f[\"ribozyme_control\"]+f[\"trnaseq\"]\n",
    "print (\"Unique ribo/trna_combinations: \"+str(len(f[\"comb_seq\"].unique())))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 159,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1str=f.to_csv(sep=\"\\t\",index=False)\n",
    "f= open(\"comb_trna_seq_plus_ribo_fix_march.csv\",\"w+\")\n",
    "f.write(df1str)\n",
    "f.close()"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
