{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#select all samples which contain the keyword phage\n",
    "#/mnt/mnemo4/microbeatlas/mapdata/20210104map1/samples.info.keywords \n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "#inp=\"/mnt/gaia/jfmrod/gaia/usr/bin/mhpc-clust -t samples -nthreads 30 -dfunc braycurtis -makecluststats /mnt/mnemo4/microbeatlas/mapdata/20200307/otutables/samples-otus.90.mapped.gz /mnt/mnemo6/lukas/Python/all_otu_samplelist/all_99_otus_samples_v0.2.2_.csv /mnt/mnemo6/lukas/Python/evo_treedists/Chloroflexi/Chloroflexi_pairs.tbl >>/mnt/mnemo6/lukas/Python/evo_treedists/Chloroflexi/Chloroflexi_BC_2.tbl\"\n",
    "\n",
    "#inp=\"/mnt/gaia/jfmrod/gaia/usr/bin/mhpc-clust -t samples -nthreads 30 -dfunc braycurtis -makecluststats /mnt/mnemo4/microbeatlas/mapdata/20200307/otutables/samples-otus.90.mapped.gz /mnt/mnemo6/lukas/Python/all_otu_samplelist/all_99_otus_samples_v0.2.2_.csv /mnt/mnemo6/lukas/Python/evo_treedists/\"+taxname+\"/\"+taxname+\"_pairs.tbl >>/mnt/mnemo6/lukas/Python/evo_treedists/\"+taxname+\"/\"+taxname+\"_BC.tbl\"\n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "grep phage /mnt/mnemo4/microbeatlas/mapdata/20210104map1/samples.info.keywords \n"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "example file: /mnt/mnemo7/jfmrod/databases.mnemo4/databases/ncbi/download/sra/sra-instant/reads/ByRun/sra/SRR/SRR975/SRR975391/SRR975391.fa.gz"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "import h5py\n",
    "import numpy as np\n",
    "import pandas as pd\n",
    "from scipy.sparse import csc_matrix\n",
    "\n",
    "def read_mapdb_h5(h5_path, otu_data=True, meta_data=True):\n",
    "    \"\"\"Example file: /mnt/mnemo3/janko/data/microbe_atlas/sruns-otus.97.otutable_plusMetaNoHeader_taxonomy_unmapped.h5\"\"\"\n",
    "    f = h5py.File(h5_path, 'r')\n",
    "    result_dict = {}\n",
    "\n",
    "    # read otu data\n",
    "    if otu_data:\n",
    "        data_handle = f[\"otu_table\"]\n",
    "        col_ptr, nzvals, rowindices, otu_index, sample_index = [np.array(data_handle[sub_key]) for sub_key in ['data_colptr', 'data_nzval', 'data_rowval', 'oids', 'sids']]\n",
    "\n",
    "        ## correct indexing (julia starts at 1)\n",
    "        col_ptr -= 1\n",
    "        rowindices -= 1\n",
    "\n",
    "        otutable_sparse = csc_matrix((nzvals,rowindices,col_ptr), shape=(data_handle[\"m\"][()],data_handle[\"n\"][()]))\n",
    "        result_dict[\"otu_data\"] = {\"otu_table\": otutable_sparse, \"otu_index\": otu_index, \"sample_index\": sample_index}\n",
    "\n",
    "    # read meta data\n",
    "    if meta_data:\n",
    "        meta_handle = f[\"meta_data\"]\n",
    "        result_dict[\"meta_data\"] = {sub_key: pd.DataFrame(np.array(meta_handle[sub_key]).T) for sub_key in meta_handle.keys()}\n",
    "\n",
    "    f.close()\n",
    "\n",
    "    return result_dict\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "#select OTU level here: in This case 99%\n",
    "f2=read_mapdb_h5(\"/mnt/mnemo3/janko/data/microbe_atlas/hdf5/v0.2.2/samples-otus.97.mapped.remap.h5\")\n",
    "\n",
    "keyw=f2[\"meta_data\"][\"keywords\"]\n",
    "\n",
    "readspec=f2[\"meta_data\"][\"readspec\"]\n",
    "\n",
    "#checking all the samples which are both wgs as well as copntain phage in their keyword\n",
    "wgs_ind=keyw[keyw[3]==\"WGS\"].index\n",
    "\n",
    "\n",
    "wgs=keyw[keyw[3]==\"WGS\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "<2174192x137800 sparse matrix of type '<class 'numpy.int32'>'\n",
       "\twith 645021862 stored elements in Compressed Sparse Column format>"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "f2[\"otu_data\"][\"otu_table\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 38,
   "metadata": {},
   "outputs": [],
   "source": [
    "all_runs_with_phages=wgs[wgs[1].str.contains(\"animal\")][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#maybe do this faster by first checking something like which samples are WGS and then check if sample in WGS: etc "
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "zcat PATH/file.gz >> TMPPATH/file.fa"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#zcat to copy all files into a tmp location"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#get all wgs path\n",
    "for i in wgs[0]:\n",
    "    inp= \"grep -m 1 \"+i.split(\".\")[0]+\" /home/jfmrod/mnemo4/work/mapdb/data/all.runs.fa.map >> /mnt/mnemo6/lukas/Python/ribozymes/version4/wgs_sample_paths.txt\"\n",
    "\n",
    "    subprocess.run(inp, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "f=open(\"gut_sample_paths.txt\")\n",
    "f_lin=f.readlines()\n",
    "\n",
    "for i in range(len(f_lin)):\n",
    "    f_lin[i]=f_lin[i].strip()\n",
    "\n",
    "#get the total length_ split it into 20 parts?\n",
    "length_total=len(f_lin)\n",
    "\n",
    "length_part=int(length_total/60)\n",
    "\n",
    "#global_fasta=\"\"\n",
    "def open_fils(filename,number):\n",
    "    #where to save the found sequence locations, to later grep out the fasta stuff\n",
    "    #or do directly in bash via grep, whats better?\n",
    "    #global global_fasta\n",
    "    #locations=[]\n",
    "    \n",
    "    filepath=\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/rnarobo_results/\"+str(number)+\"/\"+filename+\"_rnarobo_result.txt\"\n",
    "    filepath_fasta=\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/\"+str(number)+\"/\"+filename+\".fa\"\n",
    "    f1=open(filepath)\n",
    "    f_l1=f1.readlines()\n",
    "    f1.close()\n",
    "    #going thorugh the results in one file: check for start of found seq (\"  \"), then append the start position, end position, sequencename (important!) and sequence itself\n",
    "    #counter=0\n",
    "    #start=[]\n",
    "    #end=[]\n",
    "    seqnam=[]\n",
    "    #sequence=[]\n",
    "    #go trhough each line of rnarobo output\n",
    "    for line in f_l1:\n",
    "        if line[0:2]==\"  \":\n",
    "            line=line.strip().split()\n",
    "            #start.append(line[0])\n",
    "            #end.append(line[1])\n",
    "            seqnam.append(line[2])\n",
    "            counter=1\n",
    "        #add the sequence of the ribozyme    \n",
    "        #elif counter==1:\n",
    "            #line=line.strip()\n",
    "            #sequence.append(line)\n",
    "            #counter=0\n",
    "    #all reads where a ribozyme was found        \n",
    "    unique_seqnam=set(seqnam)\n",
    "    print(unique_seqnam)\n",
    "    #look for all the \n",
    "    sequence_saving=\"\"\n",
    "    for seqnam in unique_seqnam:\n",
    "        inp= \"sed -n '/\"+seqnam+\" /,/>/p' \"+filepath_fasta\n",
    "        l=subprocess.run(inp,shell=True,capture_output=True,text=True)\n",
    "        sequence_saving+=l.stdout\n",
    "        #global_fasta+=l.stdout\n",
    "    f2=open(\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/rnarobo_results_seqs/\"+str(number)+\"/\"+filename+\"_seqs_rnarobo_result.txt\",\"w\")\n",
    "    f2.write(sequence_saving)\n",
    "    f2.close\n",
    "        #print(l.stdout)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 21,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "#Trial to figure out how we can save the sequence in question without the wrong header in the end\n",
    "import subprocess\n",
    "sequence_saving=\"\"\n",
    "\n",
    "inp= \"sed -n '/D7ZQJ5M1:705:C5GNWACXX:5:1101:4750:2265 /,/>/p' /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/1/SRR2082149.fa\"\n",
    "l=subprocess.run(inp,shell=True,capture_output=True,text=True)\n",
    "sequence_saving+=l.stdout\n",
    "\n",
    "\">\".join(sequence_saving.split(\">\")[:-1])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 24,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args=\"sed -n '/D7ZQJ5M1:705:C5GNWACXX:5:1101:4750:2265 /,/>/p' /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/1/SRR2082149.fa\", returncode=2, stdout='', stderr=\"sed: can't read /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/1/SRR2082149.fa: No such file or directory\\n\")"
      ]
     },
     "execution_count": 24,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "l"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "for number in range(3,21):\n",
    "    \n",
    "    stri=\"\"\"\n",
    "import subprocess\n",
    "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/rnarobo_results/\"+str(\"\"\"+str(number)+\"\"\")\n",
    "subprocess.run(inp_start, shell=True)\n",
    "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/rnarobo_results_seqs/\"+str(\"\"\"+str(number)+\"\"\")\n",
    "subprocess.run(inp_start, shell=True)\n",
    "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/\"+str(\"\"\"+str(number)+\"\"\")\n",
    "subprocess.run(inp_start, shell=True)\n",
    "f=open(\"gut_sample_paths.txt\")\n",
    "f_lin=f.readlines()\n",
    "\n",
    "for i in range(len(f_lin)):\n",
    "    f_lin[i]=f_lin[i].strip()\n",
    "\n",
    "#get the total length_ split it into 20 parts?\n",
    "length_total=len(f_lin)\n",
    "\n",
    "length_part=int(length_total/60)\n",
    "\n",
    "#global_fasta=\"\"\n",
    "def open_fils(filename):\n",
    "    #where to save the found sequence locations, to later grep out the fasta stuff\n",
    "    #or do directly in bash via grep, whats better?\n",
    "    #global global_fasta\n",
    "    #locations=[]\n",
    "\n",
    "    filepath=\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/rnarobo_results/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+filename+\"_rnarobo_result.txt\"\n",
    "    filepath_fasta=\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+filename+\".fa\"\n",
    "    f1=open(filepath)\n",
    "    f_l1=f1.readlines()\n",
    "    f1.close()\n",
    "    #going thorugh the results in one file: check for start of found seq (\"  \"), then append the start position, end position, sequencename (important!) and sequence itself\n",
    "    #counter=0\n",
    "    #start=[]\n",
    "    #end=[]\n",
    "    seqnam=[]\n",
    "    #sequence=[]\n",
    "    #go trhough each line of rnarobo output\n",
    "    for line in f_l1:\n",
    "        if line[0:2]==\"  \":\n",
    "            line=line.strip().split()\n",
    "            #start.append(line[0])\n",
    "            #end.append(line[1])\n",
    "            seqnam.append(line[2])\n",
    "            counter=1\n",
    "        #add the sequence of the ribozyme    \n",
    "        #elif counter==1:\n",
    "            #line=line.strip()\n",
    "            #sequence.append(line)\n",
    "            #counter=0\n",
    "    #all reads where a ribozyme was found        \n",
    "    unique_seqnam=set(seqnam)\n",
    "    print(unique_seqnam)\n",
    "    #look for all the \n",
    "    sequence_saving=\"\"\n",
    "    for seqnam in unique_seqnam:\n",
    "        inp= \"sed -n '/\"+seqnam+\" /,/>/p' \"+filepath_fasta\n",
    "        l=subprocess.run(inp,shell=True,capture_output=True,text=True)\n",
    "        sequence_saving+=l.stdout\n",
    "        #global_fasta+=l.stdout\n",
    "    sequence_saving=\">\".join(sequence_saving.split(\">\")[:-1])\n",
    "    f2=open(\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/rnarobo_results_seqs/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+filename+\"_seqs_rnarobo_result.txt\",\"w\")\n",
    "    f2.write(sequence_saving)\n",
    "    f2.close\n",
    "    #print(l.stdout)\n",
    "for path in f_lin[length_part*\"\"\"+str(number)+\"\"\":length_part*(\"\"\"+str(number)+\"\"\"+1)]:\n",
    "\n",
    "    #extract the file to a new location, unzip it\n",
    "    name=path.split(\"/\")[-1].split(\".\")[0]\n",
    "    ending=path.split(\"/\")[-1].split(\".\")[-1]\n",
    "    if ending==\"gz\":\n",
    "        inp= \"zcat \"+path+\" >> /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa\"\n",
    "    elif ending==\"bz2\":\n",
    "        #dc to write to stdout and keep original file\n",
    "        inp= \"bzip2 -dc \"+path+\">> /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa\"\n",
    "    else:\n",
    "        inp=\"\"\n",
    "    #run the decompression, depoedning on which ending we have    \n",
    "    subprocess.run(inp, shell=True)\n",
    "\n",
    "    #run rnarobo, with --nratio do exclude the bad quality baseppairs\n",
    "    inp2= \"/mnt/mnemo6/lukas/anaconda3/bin/rnarobo-2.1.0-linux64 -c --nratio 0.1 /mnt/mnemo6/lukas/Python/ribozymes/version3/motif2_precise.txt /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa >> /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/rnarobo_results/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"_rnarobo_result.txt\"\n",
    "    subprocess.run(inp2, shell=True)\n",
    "\n",
    "    #cutting out sequence and saving that, as well as adding it to global variable global_fasta, which then contains all potential seuqences to scan with trnascan\n",
    "    open_fils(name)\n",
    "\n",
    "    inp3= \"rm /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa\"\n",
    "    subprocess.run(inp3, shell=True)\n",
    "    #sed through the results and cut out interesting parts, save to file? then run trnascan, or in the end?\n",
    "\"\"\"\n",
    "\n",
    "    f=open(\"test\"+str(number)+\".py\",\"w\")\n",
    "    f.write(stri)\n",
    "    f.close()\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "for number in range(1,21):\n",
      "    \n",
      "    stri=\"\"\"\n",
      "import subprocess\n",
      "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/rnarobo_results/\"+str(\"\"\"+str(number)+\"\"\")\n",
      "subprocess.run(inp_start, shell=True)\n",
      "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/rnarobo_results_seqs/\"+str(\"\"\"+str(number)+\"\"\")\n",
      "subprocess.run(inp_start, shell=True)\n",
      "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/\"+str(\"\"\"+str(number)+\"\"\")\n",
      "subprocess.run(inp_start, shell=True)\n",
      "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/trnascan/\"+str(\"\"\"+str(number)+\"\"\")\n",
      "subprocess.run(inp_start, shell=True)\n",
      "f=open(\"gut_sample_paths.txt\")\n",
      "f_lin=f.readlines()\n",
      "\n",
      "for i in range(len(f_lin)):\n",
      "    f_lin[i]=f_lin[i].strip()\n",
      "\n",
      "#get the total length_ split it into 20 parts?\n",
      "length_total=len(f_lin)\n",
      "\n",
      "length_part=int(length_total/60)\n",
      "\n",
      "#global_fasta=\"\"\n",
      "def open_fils(filename):\n",
      "    #where to save the found sequence locations, to later grep out the fasta stuff\n",
      "    #or do directly in bash via grep, whats better?\n",
      "    #global global_fasta\n",
      "    #locations=[]\n",
      "\n",
      "    filepath=\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/rnarobo_results/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+filename+\"_rnarobo_result.txt\"\n",
      "    filepath_fasta=\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+filename+\".fa\"\n",
      "    f1=open(filepath)\n",
      "    f_l1=f1.readlines()\n",
      "    f1.close()\n",
      "    #going thorugh the results in one file: check for start of found seq (\"  \"), then append the start position, end position, sequencename (important!) and sequence itself\n",
      "    #counter=0\n",
      "    #start=[]\n",
      "    #end=[]\n",
      "    seqnam=[]\n",
      "    #sequence=[]\n",
      "    #go trhough each line of rnarobo output\n",
      "    for line in f_l1:\n",
      "        if line[0:2]==\"  \":\n",
      "            line=line.strip().split()\n",
      "            #start.append(line[0])\n",
      "            #end.append(line[1])\n",
      "            seqnam.append(line[2])\n",
      "            counter=1\n",
      "        #add the sequence of the ribozyme    \n",
      "        #elif counter==1:\n",
      "            #line=line.strip()\n",
      "            #sequence.append(line)\n",
      "            #counter=0\n",
      "    #all reads where a ribozyme was found        \n",
      "    unique_seqnam=set(seqnam)\n",
      "    print(unique_seqnam)\n",
      "    #look for all the \n",
      "    sequence_saving=\"\"\n",
      "    for seqnam in unique_seqnam:\n",
      "        inp= \"sed -n '/\"+seqnam+\" /,/>/p' \"+filepath_fasta\n",
      "        l=subprocess.run(inp,shell=True,capture_output=True,text=True)\n",
      "        if len(l.stdout.split(\">\")) > 2:\n",
      "            sequence_saving+=\">\".join(l.stdout.split(\">\")[:-1])\n",
      "                    #print(seqnam)\n",
      "        else:\n",
      "            sequence_saving+=l.stdout\n",
      "    f2=open(\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/rnarobo_results_seqs/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+filename+\"_seqs_rnarobo_result.txt\",\"w\")\n",
      "\n",
      "    #sequence_saving2=\">\".join(sequence_saving.split(\">\")[:-1])\n",
      "    f2.write(sequence_saving)\n",
      "    f2.close\n",
      "    #print(l.stdout)\n",
      "for path in f_lin[length_part*\"\"\"+str(number)+\"\"\":length_part*(\"\"\"+str(number)+\"\"\"+1)]:\n",
      "\n",
      "    #extract the file to a new location, unzip it\n",
      "    name=path.split(\"/\")[-1].split(\".\")[0]\n",
      "    ending=path.split(\"/\")[-1].split(\".\")[-1]\n",
      "    if ending==\"gz\":\n",
      "        inp= \"zcat \"+path+\" >> /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa\"\n",
      "    elif ending==\"bz2\":\n",
      "        #dc to write to stdout and keep original file\n",
      "        inp= \"bzip2 -dc \"+path+\">> /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa\"\n",
      "    else:\n",
      "        inp=\"\"\n",
      "    #run the decompression, depoedning on which ending we have    \n",
      "    subprocess.run(inp, shell=True)\n",
      "\n",
      "    #run rnarobo, with --nratio do exclude the bad quality baseppairs\n",
      "    inp2= \"/mnt/mnemo6/lukas/anaconda3/bin/rnarobo-2.1.0-linux64 -c --nratio 0.1 /mnt/mnemo6/lukas/Python/ribozymes/version3/motif2_precise_T.txt /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa >> /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/rnarobo_results/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"_rnarobo_result.txt\"\n",
      "    subprocess.run(inp2, shell=True)\n",
      "\n",
      "    #cutting out sequence and saving that, as well as adding it to global variable global_fasta, which then contains all potential seuqences to scan with trnascan\n",
      "    open_fils(name)\n",
      "\n",
      "    inp3= \"rm /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa\"\n",
      "    subprocess.run(inp3, shell=True)\n",
      "    #sed through the results and cut out interesting parts, save to file? then run trnascan, or in the end?\n",
      "    inp4=\"/mnt/mnemo6/lukas/anaconda3/envs/sar3/bin/tRNAscan-SE -G --detail /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/rnarobo_results_seqs/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"_seqs_rnarobo_result.txt -o /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/trnascan/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"trna_result.txt -f /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/trnascan/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"trna_secondary.tbl\"\n",
      "    subprocess.run(inp4, shell=True)\n",
      "\"\"\"\n",
      "\n",
      "    f=open(\"test\"+str(number)+\".py\",\"w\")\n",
      "    f.write(stri)\n",
      "    f.close()\n",
      "\n"
     ]
    }
   ],
   "source": [
    "#we are replacing the original script and insetad saving it into v3_t- plus, afterwards we are also including the trna scan here\n",
    "stringa='''for number in range(1,21):\n",
    "    \n",
    "    stri=\"\"\"\n",
    "import subprocess\n",
    "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/rnarobo_results/\"+str(\"\"\"+str(number)+\"\"\")\n",
    "subprocess.run(inp_start, shell=True)\n",
    "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/rnarobo_results_seqs/\"+str(\"\"\"+str(number)+\"\"\")\n",
    "subprocess.run(inp_start, shell=True)\n",
    "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/\"+str(\"\"\"+str(number)+\"\"\")\n",
    "subprocess.run(inp_start, shell=True)\n",
    "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/trnascan/\"+str(\"\"\"+str(number)+\"\"\")\n",
    "subprocess.run(inp_start, shell=True)\n",
    "f=open(\"gut_sample_paths.txt\")\n",
    "f_lin=f.readlines()\n",
    "\n",
    "for i in range(len(f_lin)):\n",
    "    f_lin[i]=f_lin[i].strip()\n",
    "\n",
    "#get the total length_ split it into 20 parts?\n",
    "length_total=len(f_lin)\n",
    "\n",
    "length_part=int(length_total/60)\n",
    "\n",
    "#global_fasta=\"\"\n",
    "def open_fils(filename):\n",
    "    #where to save the found sequence locations, to later grep out the fasta stuff\n",
    "    #or do directly in bash via grep, whats better?\n",
    "    #global global_fasta\n",
    "    #locations=[]\n",
    "\n",
    "    filepath=\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/rnarobo_results/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+filename+\"_rnarobo_result.txt\"\n",
    "    filepath_fasta=\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+filename+\".fa\"\n",
    "    f1=open(filepath)\n",
    "    f_l1=f1.readlines()\n",
    "    f1.close()\n",
    "    #going thorugh the results in one file: check for start of found seq (\"  \"), then append the start position, end position, sequencename (important!) and sequence itself\n",
    "    #counter=0\n",
    "    #start=[]\n",
    "    #end=[]\n",
    "    seqnam=[]\n",
    "    #sequence=[]\n",
    "    #go trhough each line of rnarobo output\n",
    "    for line in f_l1:\n",
    "        if line[0:2]==\"  \":\n",
    "            line=line.strip().split()\n",
    "            #start.append(line[0])\n",
    "            #end.append(line[1])\n",
    "            seqnam.append(line[2])\n",
    "            counter=1\n",
    "        #add the sequence of the ribozyme    \n",
    "        #elif counter==1:\n",
    "            #line=line.strip()\n",
    "            #sequence.append(line)\n",
    "            #counter=0\n",
    "    #all reads where a ribozyme was found        \n",
    "    unique_seqnam=set(seqnam)\n",
    "    print(unique_seqnam)\n",
    "    #look for all the \n",
    "    sequence_saving=\"\"\n",
    "    for seqnam in unique_seqnam:\n",
    "        inp= \"sed -n '/\"+seqnam+\" /,/>/p' \"+filepath_fasta\n",
    "        l=subprocess.run(inp,shell=True,capture_output=True,text=True)\n",
    "        if len(l.stdout.split(\">\")) > 2:\n",
    "            sequence_saving+=\">\".join(l.stdout.split(\">\")[:-1])\n",
    "                    #print(seqnam)\n",
    "        else:\n",
    "            sequence_saving+=l.stdout\n",
    "    f2=open(\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/rnarobo_results_seqs/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+filename+\"_seqs_rnarobo_result.txt\",\"w\")\n",
    "\n",
    "    #sequence_saving2=\">\".join(sequence_saving.split(\">\")[:-1])\n",
    "    f2.write(sequence_saving)\n",
    "    f2.close\n",
    "    #print(l.stdout)\n",
    "for path in f_lin[length_part*\"\"\"+str(number)+\"\"\":length_part*(\"\"\"+str(number)+\"\"\"+1)]:\n",
    "\n",
    "    #extract the file to a new location, unzip it\n",
    "    name=path.split(\"/\")[-1].split(\".\")[0]\n",
    "    ending=path.split(\"/\")[-1].split(\".\")[-1]\n",
    "    if ending==\"gz\":\n",
    "        inp= \"zcat \"+path+\" >> /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa\"\n",
    "    elif ending==\"bz2\":\n",
    "        #dc to write to stdout and keep original file\n",
    "        inp= \"bzip2 -dc \"+path+\">> /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa\"\n",
    "    else:\n",
    "        inp=\"\"\n",
    "    #run the decompression, depoedning on which ending we have    \n",
    "    subprocess.run(inp, shell=True)\n",
    "\n",
    "    #run rnarobo, with --nratio do exclude the bad quality baseppairs\n",
    "    inp2= \"/mnt/mnemo6/lukas/anaconda3/bin/rnarobo-2.1.0-linux64 -c --nratio 0.1 /mnt/mnemo6/lukas/Python/ribozymes/version3/motif2_precise_T.txt /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa >> /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/rnarobo_results/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"_rnarobo_result.txt\"\n",
    "    subprocess.run(inp2, shell=True)\n",
    "\n",
    "    #cutting out sequence and saving that, as well as adding it to global variable global_fasta, which then contains all potential seuqences to scan with trnascan\n",
    "    open_fils(name)\n",
    "\n",
    "    inp3= \"rm /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa\"\n",
    "    subprocess.run(inp3, shell=True)\n",
    "    #sed through the results and cut out interesting parts, save to file? then run trnascan, or in the end?\n",
    "    inp4=\"/mnt/mnemo6/lukas/anaconda3/envs/sar3/bin/tRNAscan-SE -G --detail /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/rnarobo_results_seqs/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"_seqs_rnarobo_result.txt -o /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/trnascan/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"trna_result.txt -f /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v3_T/trnascan/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"trna_secondary.tbl\"\n",
    "    subprocess.run(inp4, shell=True)\n",
    "\"\"\"\n",
    "\n",
    "    f=open(\"test\"+str(number)+\".py\",\"w\")\n",
    "    f.write(stri)\n",
    "    f.close()\n",
    "'''\n",
    "\n",
    "print(stringa.replace(\"v3_T\",\"v4\"))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "for number in range(1,61):\n",
    "    \n",
    "    stri=\"\"\"\n",
    "import subprocess\n",
    "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/rnarobo_results/\"+str(\"\"\"+str(number)+\"\"\")\n",
    "subprocess.run(inp_start, shell=True)\n",
    "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/rnarobo_results_seqs/\"+str(\"\"\"+str(number)+\"\"\")\n",
    "subprocess.run(inp_start, shell=True)\n",
    "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/\"+str(\"\"\"+str(number)+\"\"\")\n",
    "subprocess.run(inp_start, shell=True)\n",
    "inp_start=\"mkdir /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/trnascan/\"+str(\"\"\"+str(number)+\"\"\")\n",
    "subprocess.run(inp_start, shell=True)\n",
    "f=open(\"wgs_sample_paths.txt\")\n",
    "f_lin=f.readlines()\n",
    "\n",
    "for i in range(len(f_lin)):\n",
    "    f_lin[i]=f_lin[i].strip()\n",
    "\n",
    "#get the total length_ split it into 60 parts?\n",
    "length_total=len(f_lin)\n",
    "\n",
    "length_part=int(length_total/60)\n",
    "\n",
    "#global_fasta=\"\"\n",
    "def open_fils(filename):\n",
    "    #where to save the found sequence locations, to later grep out the fasta stuff\n",
    "    #or do directly in bash via grep, whats better?\n",
    "    #global global_fasta\n",
    "    #locations=[]\n",
    "\n",
    "    filepath=\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/rnarobo_results/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+filename+\"_rnarobo_result.txt\"\n",
    "    filepath_fasta=\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+filename+\".fa\"\n",
    "    f1=open(filepath)\n",
    "    f_l1=f1.readlines()\n",
    "    f1.close()\n",
    "    #going thorugh the results in one file: check for start of found seq (\"  \"), then append the start position, end position, sequencename (important!) and sequence itself\n",
    "    #counter=0\n",
    "    #start=[]\n",
    "    #end=[]\n",
    "    seqnam=[]\n",
    "    #sequence=[]\n",
    "    #go trhough each line of rnarobo output\n",
    "    for line in f_l1:\n",
    "        if line[0:2]==\"  \":\n",
    "            line=line.strip().split()\n",
    "            #start.append(line[0])\n",
    "            #end.append(line[1])\n",
    "            seqnam.append(line[2])\n",
    "            counter=1\n",
    "        #add the sequence of the ribozyme    \n",
    "        #elif counter==1:\n",
    "            #line=line.strip()\n",
    "            #sequence.append(line)\n",
    "            #counter=0\n",
    "    #all reads where a ribozyme was found        \n",
    "    unique_seqnam=set(seqnam)\n",
    "    print(unique_seqnam)\n",
    "    #look for all the \n",
    "    sequence_saving=\"\"\n",
    "    for seqnam in unique_seqnam:\n",
    "        inp= \"sed -n '/\"+seqnam+\" /,/>/p' \"+filepath_fasta\n",
    "        l=subprocess.run(inp,shell=True,capture_output=True,text=True)\n",
    "        if len(l.stdout.split(\">\")) > 2:\n",
    "            sequence_saving+=\">\".join(l.stdout.split(\">\")[:-1])\n",
    "                    #print(seqnam)\n",
    "        else:\n",
    "            sequence_saving+=l.stdout\n",
    "    f2=open(\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/rnarobo_results_seqs/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+filename+\"_seqs_rnarobo_result.txt\",\"w\")\n",
    "\n",
    "    #sequence_saving2=\">\".join(sequence_saving.split(\">\")[:-1])\n",
    "    f2.write(sequence_saving)\n",
    "    f2.close\n",
    "    #print(l.stdout)\n",
    "for path in f_lin[length_part*\"\"\"+str(number)+\"\"\":length_part*(\"\"\"+str(number)+\"\"\"+1)]:\n",
    "\n",
    "    #extract the file to a new location, unzip it\n",
    "    name=path.split(\"/\")[-1].split(\".\")[0]\n",
    "    ending=path.split(\"/\")[-1].split(\".\")[-1]\n",
    "    if ending==\"gz\":\n",
    "        inp= \"zcat \"+path+\" >> /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa\"\n",
    "    elif ending==\"bz2\":\n",
    "        #dc to write to stdout and keep original file\n",
    "        inp= \"bzip2 -dc \"+path+\">> /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa\"\n",
    "    else:\n",
    "        inp=\"\"\n",
    "    #run the decompression, depoedning on which ending we have    \n",
    "    subprocess.run(inp, shell=True)\n",
    "\n",
    "    #run rnarobo, with --nratio do exclude the bad quality baseppairs\n",
    "    inp2= \"/mnt/mnemo6/lukas/anaconda3/bin/rnarobo-2.1.0-linux64 -c --nratio 0.1 /mnt/mnemo6/lukas/Python/ribozymes/version4/motif2_precise.txt /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa >> /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/rnarobo_results/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"_rnarobo_result.txt\"\n",
    "    subprocess.run(inp2, shell=True)\n",
    "\n",
    "    #cutting out sequence and saving that, as well as adding it to global variable global_fasta, which then contains all potential seuqences to scan with trnascan\n",
    "    open_fils(name)\n",
    "\n",
    "    inp3= \"rm /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\".fa\"\n",
    "    subprocess.run(inp3, shell=True)\n",
    "    #sed through the results and cut out interesting parts, save to file? then run trnascan, or in the end?\n",
    "    inp4=\"/mnt/mnemo6/lukas/anaconda3/envs/sar3/bin/tRNAscan-SE -G --detail /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/rnarobo_results_seqs/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"_seqs_rnarobo_result.txt -o /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/trnascan/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"trna_result.txt -f /mnt/mnemo6/lukas/Python/ribozymes/phage_samples/v4/trnascan/\"+str(\"\"\"+str(number)+\"\"\")+\"/\"+name+\"trna_secondary.tbl\"\n",
    "    subprocess.run(inp4, shell=True)\n",
    "\"\"\"\n",
    "\n",
    "    f=open(\"test\"+str(number)+\".py\",\"w\")\n",
    "    f.write(stri)\n",
    "    f.close()\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\"\"\"\n",
    "com_fas=global_fasta.split(\"\\n\")\n",
    "\n",
    "fasta_corr=[]\n",
    "checker=0\n",
    "fasta_part=\"\"\n",
    "#to -1 because last one is empty for whatever reason\n",
    "for line in com_fas[:-1]:\n",
    "    \n",
    "    if line[0]==\">\":\n",
    "        \n",
    " \n",
    "        if checker==0:\n",
    "            fasta_corr.append(fasta_part)\n",
    "            fasta_part=\"\"\n",
    "            line=line.split()\n",
    "            fasta_corr.append(line[0])\n",
    "            checker=1\n",
    "        else:\n",
    "            checker=0\n",
    "    else:\n",
    "        fasta_part+=line.strip()\n",
    "        \n",
    "\n",
    "fasta_corr=fasta_corr[1:]\n",
    "\n",
    "fast_seq=\"\\n\".join(fasta_corr)\n",
    "\n",
    "fast_seq=fast_seq.replace(\".\",\"_\")\n",
    "\n",
    "f= open(\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/new/trna_scan/all_sequences.fasta\",\"w+\")\n",
    "f.write(fast_seq)\n",
    "f.close()\n",
    "#trnascan here\n",
    "\"\"\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "''"
      ]
     },
     "execution_count": 36,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "global_fasta"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 42,
   "metadata": {},
   "outputs": [],
   "source": [
    "rs_1=open(\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/new/trna_scan/trnascan_results.txt\")\n",
    "rs_s=rs_1.readlines()\n",
    "rs_1.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create dict for trna beginnings\n",
    "dicti={}\n",
    "for line in rs_s[4:]:\n",
    "    #create a dictionary with seq name as key and the starting position as value\n",
    "        line=line.strip().split()\n",
    "        if line[0] in dicti:\n",
    "            dicti[line[0]].append(int(line[3]))\n",
    "        else:\n",
    "            dicti[line[0]]=[int(line[3])]\n",
    "            \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "metadata": {},
   "outputs": [],
   "source": [
    "#lists all the files in the directory\n",
    "filenames_filt=[]\n",
    "filenames=os.listdir(\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/new/rnarobo_results/\")\n",
    "for fi in filenames:\n",
    "    if fi[-3:]==\"txt\":\n",
    "        filenames_filt.append(fi)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "metadata": {},
   "outputs": [],
   "source": [
    "#open multiple results files and saving to dict\n",
    "dicti_r={}\n",
    "import os\n",
    "for filename in filenames_filt:\n",
    "   with open(os.path.join(\"/mnt/mnemo6/lukas/Python/ribozymes/phage_samples/new/rnarobo_results/\", filename), 'r') as f: # open in readonly mode\n",
    "      # do your stuff\n",
    "        f_l=f.readlines()\n",
    "        #means at least one ribo was found\n",
    "        if len(f_l)>17:\n",
    "            for linea in f_l[11:-6]:\n",
    "                if linea[0]!=\"|\":\n",
    "                    line=linea.replace(\".\",\"_\")\n",
    "                    line=line.strip().split()\n",
    "                    if line[2] in dicti_r:\n",
    "                        dicti_r[line[2]].append(int(line[0]))\n",
    "                    else:\n",
    "                        dicti_r[line[2]]=[int(line[0])]\n",
    "\n",
    "            \n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "7533\n"
     ]
    }
   ],
   "source": [
    "print(sum([len(dicti_r[x]) for x in dicti_r if isinstance(dicti_r[x], list)]))"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 71,
   "metadata": {},
   "outputs": [],
   "source": [
    "matching=[]\n",
    "for key in dicti_r:\n",
    "    if key in dicti:\n",
    "        for value in dicti_r[key]:\n",
    "            for value2 in dicti[key]:\n",
    "                if value-value2 in range(-15,15):\n",
    "                    matching.append([key,value,value2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 98,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['SRR5088933_4044330', 189, 177],\n",
       " ['SRR6379761_193842', 69, 70],\n",
       " ['SRR6379761_839598', 69, 67],\n",
       " ['SRR6379761_6054261', 229, 231],\n",
       " ['SRR6379761_7115974', 218, 219],\n",
       " ['SRR6379761_7990996', 69, 67],\n",
       " ['SRR6379761_9807507', 244, 243],\n",
       " ['SRR6379761_16636702', 205, 206],\n",
       " ['SRR6379761_17261682', 223, 231],\n",
       " ['SRR6379761_19449059', 223, 231],\n",
       " ['SRR6379761_20325740', 223, 231],\n",
       " ['SRR6379761_20502096', 75, 80],\n",
       " ['SRR6379761_21551329', 214, 212],\n",
       " ['SRR6379761_21880813', 214, 212],\n",
       " ['SRR6379761_23140049', 218, 219],\n",
       " ['SRR6379761_25657473', 69, 67],\n",
       " ['SRR6379761_29330552', 69, 67],\n",
       " ['SRR6379761_33382391', 69, 67]]"
      ]
     },
     "execution_count": 98,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matching"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 79,
   "metadata": {},
   "outputs": [],
   "source": [
    "#blast result\n",
    "\n",
    "blstr=pd.read_csv(\"/mnt/mnemo6/lukas/Python/ribozymes/trna_database/test_3.tbl\",sep=\"\\t\",header=None)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 84,
   "metadata": {},
   "outputs": [],
   "source": [
    "blstr.columns=[\"qseqid\", \"sseqid\", \"pident\", \"length\", \"mismatch\", \"gapopen\", \"qstart\" ,\"qend\", \"sstart\" ,\"send\", \"evalue\", \"bitscore\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "metadata": {},
   "outputs": [],
   "source": [
    "dicti_t_2={}\n",
    "for ids in range(len(blstr)):\n",
    "    #create a dictionary with seq name as key and the starting position as value\n",
    "        if blstr.iloc[ids,0] in dicti_t_2:\n",
    "            dicti_t_2[blstr.iloc[ids,0]].append(int(blstr.iloc[ids,7]))\n",
    "        else:\n",
    "            dicti_t_2[blstr.iloc[ids,0]]=[int(blstr.iloc[ids,7])]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 93,
   "metadata": {},
   "outputs": [],
   "source": [
    "matching_2=[]\n",
    "for key in dicti_r:\n",
    "    if key in dicti_t_2:\n",
    "        for value in dicti_r[key]:\n",
    "            for value2 in dicti_t_2[key]:\n",
    "                if value-value2 in range(-15,15):\n",
    "                    matching_2.append([key,value,value2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 94,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "[['SRR5088933_847575', 142, 139],\n",
       " ['SRR5088933_847575', 142, 139],\n",
       " ['SRR5088933_2776739', 144, 141],\n",
       " ['SRR5088933_2776739', 144, 141],\n",
       " ['SRR5088933_3167187', 45, 42],\n",
       " ['SRR5088933_3167187', 45, 42],\n",
       " ['SRR5088933_3817122', 138, 135],\n",
       " ['SRR5088933_3817122', 138, 135],\n",
       " ['SRR5088933_4004101', 34, 31],\n",
       " ['SRR5088933_4004101', 34, 31],\n",
       " ['SRR5088933_5160924', 38, 35],\n",
       " ['SRR5088933_5160924', 38, 35],\n",
       " ['SRR5088933_7329507', 143, 140],\n",
       " ['SRR5088933_7329507', 143, 140],\n",
       " ['SRR5088933_7743988', 147, 144],\n",
       " ['SRR5088933_7743988', 147, 144],\n",
       " ['SRR5088933_7757114', 139, 136],\n",
       " ['SRR5088933_7757114', 139, 136],\n",
       " ['SRR5088933_7769075', 48, 45],\n",
       " ['SRR5088933_7769075', 48, 45],\n",
       " ['SRR5088933_8052701', 34, 31],\n",
       " ['SRR5088933_8052701', 34, 31],\n",
       " ['SRR5088933_8145875', 40, 37],\n",
       " ['SRR5088933_8145875', 40, 37],\n",
       " ['SRR5088933_8166443', 145, 142],\n",
       " ['SRR5088933_8166443', 145, 142],\n",
       " ['SRR5088933_8191347', 42, 39],\n",
       " ['SRR5088933_8191347', 42, 39],\n",
       " ['SRR5088933_8313991', 32, 29],\n",
       " ['SRR5088933_8313991', 32, 29],\n",
       " ['SRR5088933_8644761', 38, 35],\n",
       " ['SRR5088933_8644761', 38, 35],\n",
       " ['SRR5088933_8726033', 139, 136],\n",
       " ['SRR5088933_8726033', 139, 136],\n",
       " ['SRR5088933_8789347', 46, 43],\n",
       " ['SRR5088933_8789347', 46, 43],\n",
       " ['SRR5088933_8883562', 147, 144],\n",
       " ['SRR5088933_8883562', 147, 144],\n",
       " ['SRR5088933_9021817', 38, 35],\n",
       " ['SRR5088933_9021817', 38, 35],\n",
       " ['SRR5088933_9187782', 139, 136],\n",
       " ['SRR5088933_9187782', 139, 136],\n",
       " ['SRR5088933_9275244', 38, 35],\n",
       " ['SRR5088933_9275244', 38, 35],\n",
       " ['SRR5088933_9640200', 139, 136],\n",
       " ['SRR5088933_9640200', 139, 136],\n",
       " ['SRR5088933_9807304', 46, 43],\n",
       " ['SRR5088933_9807304', 46, 43],\n",
       " ['SRR5088933_9807305', 46, 43],\n",
       " ['SRR5088933_9807305', 46, 43],\n",
       " ['SRR5088933_9807306', 46, 43],\n",
       " ['SRR5088933_9807306', 46, 43],\n",
       " ['SRR5088933_9845776', 147, 144],\n",
       " ['SRR5088933_9845776', 147, 144],\n",
       " ['SRR5088933_9881693', 139, 136],\n",
       " ['SRR5088933_9881693', 139, 136],\n",
       " ['SRR5088933_9957621', 46, 43],\n",
       " ['SRR5088933_9957621', 46, 43],\n",
       " ['SRR5088933_9977553', 42, 39],\n",
       " ['SRR5088933_9977553', 42, 39],\n",
       " ['SRR5088933_10098951', 42, 39],\n",
       " ['SRR5088933_10098951', 42, 39],\n",
       " ['SRR5088933_10139149', 139, 136],\n",
       " ['SRR5088933_10139149', 139, 136],\n",
       " ['SRR5088933_10456044', 38, 35],\n",
       " ['SRR5088933_10456044', 38, 35],\n",
       " ['SRR5088933_10465121', 139, 136],\n",
       " ['SRR5088933_10465121', 139, 136],\n",
       " ['SRR5088933_11036133', 139, 136],\n",
       " ['SRR5088933_11036133', 139, 136],\n",
       " ['SRR5088933_11057447', 139, 136],\n",
       " ['SRR5088933_11057447', 139, 136],\n",
       " ['SRR5088933_11168567', 38, 35],\n",
       " ['SRR5088933_11168567', 38, 35],\n",
       " ['SRR061135_7034162', 110, 100],\n",
       " ['SRR061135_7034162', 110, 100],\n",
       " ['SRR061135_7034162', 110, 100],\n",
       " ['SRR061135_7034162', 110, 100],\n",
       " ['SRR061135_7034162', 110, 100],\n",
       " ['SRR061135_7034162', 110, 100],\n",
       " ['SRR061135_7034162', 110, 100],\n",
       " ['SRR061135_7034162', 110, 100],\n",
       " ['SRR061135_7034162', 110, 100],\n",
       " ['SRR6379761_1891413', 213, 210],\n",
       " ['SRR6379761_1891413', 213, 210],\n",
       " ['SRR6379761_6054261', 229, 226],\n",
       " ['SRR6379761_6054261', 229, 226],\n",
       " ['SRR6379761_8578583', 188, 185],\n",
       " ['SRR6379761_8578583', 188, 185],\n",
       " ['SRR6379761_9807507', 244, 241],\n",
       " ['SRR6379761_9807507', 244, 241],\n",
       " ['SRR6379761_20666756', 37, 34],\n",
       " ['SRR6379761_20666756', 37, 34],\n",
       " ['SRR6379761_28033759', 37, 34],\n",
       " ['SRR6379761_28033759', 37, 34],\n",
       " ['SRR1804201_15293458', 110, 100],\n",
       " ['SRR1804201_15293458', 110, 100],\n",
       " ['SRR1804201_15293458', 110, 100],\n",
       " ['SRR1804201_15293458', 110, 100],\n",
       " ['SRR1804201_15293458', 110, 100],\n",
       " ['SRR1804201_15293458', 110, 100],\n",
       " ['SRR1804201_15293458', 110, 100],\n",
       " ['SRR1804201_15293458', 110, 100],\n",
       " ['SRR1804201_15293458', 110, 100]]"
      ]
     },
     "execution_count": 94,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matching_2"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#maybe parellelizing? \n",
    "import subprocess\n",
    "\n",
    "template = 'python script.py {} {} {} {}'\n",
    "\n",
    "args = [[1, 2, 3, 4], [5, 6, 7, 8]]\n",
    "\n",
    "# Run commands in parallel\n",
    "processes = []\n",
    "\n",
    "for arg in args:\n",
    "    command = template.format(*[str(a) for a in arg])\n",
    "    process = subprocess.Popen(command, shell=True)\n",
    "    processes.append(process)\n",
    "\n",
    "# Collect statuses\n",
    "output = [p.wait() for p in processes]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
