{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 2,
   "metadata": {},
   "outputs": [],
   "source": [
    "import subprocess\n",
    "import pandas as pd\n",
    "import numpy as np"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [],
   "source": [
    "inp=\"/mnt/mnemo6/lukas/anaconda3/bin/rnarobo-2.1.0-linux64 -c --nratio 0.1 /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v2_with_4/motif2_precise_v2.txt /mnt/mnemo6/lukas/Python/ribozymes/genome/combined/1_and_2_all_seqs.fasta >> /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v2_with_4/comb_rnarobo_result_precise.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#false positive with first version\n",
    "bases=[\"C\",\"T\",\"G\",\"A\"]\n",
    "for bas in bases: \n",
    "    inp=\"/mnt/mnemo6/lukas/anaconda3/bin/rnarobo-2.1.0-linux64 -c --nratio 0.1 /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/motif2_C_to_\"+bas+\".txt /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/combined.fna >> /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/comb_rnarobo_result_precise.txt\"\n",
    "    subprocess.run(inp, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "CompletedProcess(args='/mnt/mnemo6/lukas/anaconda3/bin/rnarobo-2.1.0-linux64 -c --nratio 0.1 /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v2_with_4/motif2_precise_v2.txt /mnt/mnemo6/lukas/Python/ribozymes/genome/combined/1_and_2_all_seqs.fasta >> /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v2_with_4/comb_rnarobo_result_precise.txt', returncode=0)"
      ]
     },
     "execution_count": 3,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subprocess.run(inp, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "for bas in bases: \n",
    "    complete_fasta=\"\"\n",
    "    filepath=\"/mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/comb_rnarobo_result_precise.txt\"\n",
    "    filepath_fasta=\"/mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/combined.fna\"\n",
    "    f=open(filepath)\n",
    "    f_l=f.readlines()\n",
    "    #going thorugh the results in one file: check for start of found seq (\"  \"), then append the start position, end position, sequencename (important!) and sequence itself\n",
    "    counter=0\n",
    "    #start=[]\n",
    "    #end=[]\n",
    "    seqnam=[]\n",
    "    sequence=[]\n",
    "    #go trhough each line of rnarobo output\n",
    "    for line in f_l[11:-6]:\n",
    "        #if len(line)==0:\n",
    "           # break\n",
    "        if not line[0]==\"|\":\n",
    "            line=line.strip().split()\n",
    "            #start.append(line[0])\n",
    "            #end.append(line[1])\n",
    "            seqnam.append(line[2])\n",
    "            counter=1\n",
    "        #add the sequence of the ribozyme    \n",
    "        elif counter==1:\n",
    "            line=line.strip()\n",
    "            sequence.append(line)\n",
    "            counter=0\n",
    "\n",
    "    #all reads where a ribozyme was found        \n",
    "    unique_seqnam=set(seqnam)\n",
    "    #look for all the \n",
    "    for seqname in unique_seqnam:\n",
    "        inp= \"sed -n '/\"+seqname+\"/,/>/p' \"+filepath_fasta\n",
    "        l=subprocess.run(inp,shell=True,capture_output=True,text=True)\n",
    "        complete_fasta+=l.stdout\n",
    "        #print(l.stdout)\n",
    "    com_fas=complete_fasta.split(\"\\n\")\n",
    "    fasta_corr=[]\n",
    "    checker=0\n",
    "    fasta_part=\"\"\n",
    "    #to -1 because last one is empty for whatever reason\n",
    "    for line in com_fas[:-1]:\n",
    "\n",
    "        if line==\"\": \n",
    "            continue\n",
    "\n",
    "        elif line[0]==\">\":\n",
    "\n",
    "\n",
    "            if checker==0:\n",
    "                fasta_corr.append(fasta_part)\n",
    "                fasta_part=\"\"\n",
    "                line=line.split()\n",
    "                fasta_corr.append(line[0])\n",
    "                checker=1\n",
    "            else:\n",
    "                checker=0\n",
    "\n",
    "        else:\n",
    "            fasta_part+=line.strip()\n",
    "\n",
    "\n",
    "    fasta_corr=fasta_corr[1:]\n",
    "\n",
    "    fast_seq=\"\\n\".join(fasta_corr)\n",
    "\n",
    "    fast_seq=fast_seq.replace(\".\",\"_\")\n",
    "\n",
    "\n",
    "    f= open(\"/mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_all_viral_sequences_v2.fasta\",\"w+\")\n",
    "    f.write(fast_seq)\n",
    "    f.close()    \n",
    "    \n",
    "    \n",
    "    inp2=\"/mnt/mnemo6/lukas/anaconda3/envs/sar3/bin/tRNAscan-SE -G /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_all_viral_sequences_v2.fasta >> /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_trna_scan_v2.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "bases=[\"C\",\"T\",\"G\",\"A\"]\n",
    "for bas in bases: \n",
    "    inp2=\"/mnt/mnemo6/lukas/anaconda3/envs/sar3/bin/tRNAscan-SE -G /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_all_viral_sequences_v2.fasta -o /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_trna_scan_v2.txt -f /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/falsepositive/first/\"+bas+\"/combined_trna_scan_secondary_struc.txt\"\n",
    "    subprocess.run(inp2, shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 36,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 37,
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "fasta_corr=fasta_corr[1:]\n",
    "\n",
    "fast_seq=\"\\n\".join(fasta_corr)\n",
    "\n",
    "fast_seq=fast_seq.replace(\".\",\"_\")\n",
    "\n",
    "\n",
    "f= open(\"combined_all_viral_sequences_v2.fasta\",\"w+\")\n",
    "f.write(fast_seq)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "metadata": {},
   "outputs": [],
   "source": [
    "inp2=\"/mnt/mnemo6/lukas/anaconda3/envs/sar3/bin/tRNAscan-SE -G /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/combined_all_viral_sequences_v2.fasta >> /mnt/mnemo6/lukas/Python/ribozymes/genome/combined_v3/combined_trna_scan_v2.txt\""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "'/mnt/mnemo6/lukas/Python/ribozymes/genome/combined'"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "subprocess.run(inp2,shell=True)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  },
  {
   "cell_type": "code",
   "execution_count": 15,
   "metadata": {},
   "outputs": [],
   "source": [
    "rnarobo_=open(\"comb_rnarobo_result_precise.txt\")\n",
    "rnarobo=rnarobo_.readlines()\n",
    "rnarobo_.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 16,
   "metadata": {},
   "outputs": [],
   "source": [
    "trna_=open(\"combined_trna_scan_v2.txt\")\n",
    "trna=trna_.readlines()\n",
    "trna_.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [],
   "source": [
    "#rnarobo[11:-6] gives us the part we need"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 74,
   "metadata": {},
   "outputs": [],
   "source": [
    "#trna[5:] gives us the part we need"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 17,
   "metadata": {},
   "outputs": [],
   "source": [
    "dicti={}\n",
    "for line in trna[4:]:\n",
    "    #create a dictionary with seq name as key and the starting position as value\n",
    "        line=line.strip().split()\n",
    "        if line[0] in dicti:\n",
    "            dicti[line[0]].append(int(line[3]))\n",
    "        else:\n",
    "            dicti[line[0]]=[int(line[3])]\n",
    "            \n",
    "        \n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 18,
   "metadata": {},
   "outputs": [],
   "source": [
    "dicti_r={}\n",
    "for line in rnarobo[11:-6]:\n",
    "    if line[0]!=\"|\":\n",
    "        line=line.replace(\".\",\"_\")\n",
    "        line=line.strip().split()\n",
    "        if line[2] in dicti_r:\n",
    "            dicti_r[line[2]].append(int(line[0]))\n",
    "        else:\n",
    "            dicti_r[line[2]]=[int(line[0])]\n",
    "        "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# uvig_225424 is duplicated in the trna result, but why? and why not in the rnarobo thingy?"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 19,
   "metadata": {},
   "outputs": [],
   "source": [
    "matching=[]\n",
    "for key in dicti_r:\n",
    "    if key in dicti:\n",
    "        for value in dicti_r[key]:\n",
    "            for value2 in dicti[key]:\n",
    "                if value-value2 in range(-10,10):\n",
    "                    matching.append([key,value,value2])"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 20,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "304"
      ]
     },
     "execution_count": 20,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "len(matching)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "dicti_r[\"uvig_225424\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 21,
   "metadata": {},
   "outputs": [],
   "source": [
    "#filter out duplicates, which are in for whatever reason\n",
    "newm=[]\n",
    "for mat in matching:\n",
    "    if mat not in newm:\n",
    "        newm.append(mat)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 22,
   "metadata": {},
   "outputs": [],
   "source": [
    "lets_do_this=pd.DataFrame(columns=[\"ID\",\"trna_nr\",\"trna_start\",\"trna_end\",\"trna_type\",\"codon\",\"ins_start\",\"ins_end\",\"trna_confidence\",\"psuedo\",\"rnarobo_start\",\"rnarobo_end\",\"rnarobo_seq\"])\n",
    "check=0\n",
    "duplicates=[]\n",
    "for nr in range(len(newm)):\n",
    "    match=newm[nr]\n",
    "    lin_df=\"\"\n",
    "    for line in trna[5:]:\n",
    "    #create a dictionary with seq name as key and the starting position as value\n",
    "        line=line.strip().split()\n",
    "        \n",
    "        if match[0]==line[0]:\n",
    "            if len(lin_df.split())==0:\n",
    "                if str(match[2])==line[3]:\n",
    "                    #print(line)\n",
    "                    lin_df+=\" \".join(line)+\" \"\n",
    "                    if lin_df[-2]!=\"o\":\n",
    "                        lin_df+=\"NA \"\n",
    "            else: \n",
    "                duplicates.append(line)\n",
    "    for line in rnarobo[11:-6]:\n",
    "        line=line.replace(\".\",\"_\")\n",
    "        line=line.strip().split()\n",
    "        if len(lin_df.split())<13:\n",
    "\n",
    "            if check==1:\n",
    "                #if len(lin_df.split())<13:\n",
    "                    lin_df+=str(line[0])\n",
    "                    check=0\n",
    "                #else:\n",
    "                        #duplicates.append(line)\n",
    "\n",
    "                #check=0\n",
    "            if str(match[1])==line[0]:\n",
    "                #if len(lin_df.split())<13:\n",
    "                    lin_df+=\" \".join(line[:2])\n",
    "                    lin_df+=\" \"\n",
    "                    check=1\n",
    "            #else:\n",
    "                #duplicates.append(line)\n",
    "                #check=0\n",
    "    lin_f=lin_df.split()\n",
    "    lets_do_this.loc[nr]=np.array(lin_f)\n",
    "    #lets_do_this.iloc[nr,:8]=lin_df[0]\n",
    "    #lets_do_this.iloc[nr,8:-1]=lin_df[1]\n",
    "    #lets_do_this.iloc[nr,-1]=lin_df[2]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "lets_do_this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 1,
   "metadata": {},
   "outputs": [
    {
     "ename": "NameError",
     "evalue": "name 'lets_do_this' is not defined",
     "output_type": "error",
     "traceback": [
      "\u001b[0;31m---------------------------------------------------------------------------\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m                                 Traceback (most recent call last)",
      "\u001b[0;32m<ipython-input-1-409772efdbcf>\u001b[0m in \u001b[0;36m<module>\u001b[0;34m\u001b[0m\n\u001b[0;32m----> 1\u001b[0;31m \u001b[0mlets_do_this\u001b[0m\u001b[0;34m[\u001b[0m\u001b[0;34m\"trna_type\"\u001b[0m\u001b[0;34m]\u001b[0m\u001b[0;34m.\u001b[0m\u001b[0munique\u001b[0m\u001b[0;34m(\u001b[0m\u001b[0;34m)\u001b[0m\u001b[0;34m\u001b[0m\u001b[0;34m\u001b[0m\u001b[0m\n\u001b[0m",
      "\u001b[0;31mNameError\u001b[0m: name 'lets_do_this' is not defined"
     ]
    }
   ],
   "source": [
    "lets_do_this[\"trna_type\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 27,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "Met      148\n",
       "Leu       72\n",
       "Sup       38\n",
       "Val       16\n",
       "Pro        9\n",
       "Lys        5\n",
       "Trp        4\n",
       "Ser        4\n",
       "Thr        3\n",
       "Cys        2\n",
       "Undet      1\n",
       "Glu        1\n",
       "Name: trna_type, dtype: int64"
      ]
     },
     "execution_count": 27,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "lets_do_this[\"trna_type\"].value_counts()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "fff=lets_do_this"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1str=lets_do_this.to_csv(sep=\"\\t\")\n",
    "f= open(\"comb_trna_v2_plus_ribo.csv\",\"w+\")\n",
    "f.write(df1str)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "metadata": {},
   "outputs": [],
   "source": [
    "fff=pd.read_csv(\"comb_trna_v2_plus_ribo.csv\",sep=\"\\t\",index_col=0)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 30,
   "metadata": {},
   "outputs": [],
   "source": [
    "fff[\"trnaseq\"]= \" \"\n",
    "\n",
    "fff[\"ribozyme_control\"]= \" \""
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "#lets try to save and cut out the respective trnas (and in between part)\n",
    "for ID in range(len(fff)):\n",
    "    if fff.iloc[ID,0][0]==\"u\":\n",
    "        inp=\"grep -A1 -m1 \"+fff.iloc[ID,0]+\" ../2/GPD_sequences.fa\" \n",
    "        res=subprocess.run(inp,shell=True, capture_output=True, text=True)\n",
    "        sequ=res.stdout.split()[-1]\n",
    "        start=fff.iloc[ID,2]-1\n",
    "        end=fff.iloc[ID,3]\n",
    "        start2=fff.iloc[ID,10]-1\n",
    "        end2=fff.iloc[ID,11]\n",
    "        if start < end:\n",
    "            trna=sequ[start:end]\n",
    "            rnarobo=sequ[start2:end2]\n",
    "            fff.iloc[ID,-2]=trna\n",
    "            fff.iloc[ID,-1]=rnarobo\n",
    "        if end < start:\n",
    "            #reverse sequence: add +1 to reverse previous indexing correction \n",
    "            end-=1\n",
    "            start+=1\n",
    "            end2-=1\n",
    "            start2+=1\n",
    "            trna_rev=sequ[end:start]\n",
    "            ribo_rev=sequ[end2:start2]\n",
    "            trna=[]\n",
    "            rnarobo=[]\n",
    "            #reverse parts: transcribe back into the correct format!\n",
    "            for nuc in range(len(trna_rev)-1,-1,-1):\n",
    "                if trna_rev[nuc]==\"A\":\n",
    "                    trna.append(\"T\")\n",
    "                if trna_rev[nuc]==\"T\":\n",
    "                    trna.append(\"A\")\n",
    "                if trna_rev[nuc]==\"G\":\n",
    "                    trna.append(\"C\")\n",
    "                if trna_rev[nuc]==\"C\":\n",
    "                    trna.append(\"G\")\n",
    "            for nuc in range(len(ribo_rev)-1,-1,-1):\n",
    "                if ribo_rev[nuc]==\"A\":\n",
    "                    rnarobo.append(\"T\")\n",
    "                if ribo_rev[nuc]==\"T\":\n",
    "                    rnarobo.append(\"A\")\n",
    "                if ribo_rev[nuc]==\"G\":\n",
    "                    rnarobo.append(\"C\")\n",
    "                if ribo_rev[nuc]==\"C\":\n",
    "                    rnarobo.append(\"G\")\n",
    "            trna=\"\".join(trna)\n",
    "            rnarobo=\"\".join(rnarobo)\n",
    "            fff.iloc[ID,-2]=trna\n",
    "            fff.iloc[ID,-1]=rnarobo\n",
    "    else:\n",
    "        resolved_taxname=fff.iloc[ID,0].replace(\"sp__\",\"sp._\")\n",
    "        inp=\"grep -A1 -m1 \"+resolved_taxname+\" ../1/CHVD_clustered_mash99_v1.fasta\" \n",
    "        res=subprocess.run(inp,shell=True, capture_output=True, text=True)\n",
    "        sequ=res.stdout.split()[-1]\n",
    "        start=fff.iloc[ID,2]-1\n",
    "        end=fff.iloc[ID,3]\n",
    "        start2=fff.iloc[ID,10]-1\n",
    "        end2=fff.iloc[ID,11]\n",
    "        if start < end:\n",
    "            trna=sequ[start:end]\n",
    "            rnarobo=sequ[start2:end2]\n",
    "            fff.iloc[ID,-2]=trna\n",
    "            fff.iloc[ID,-1]=rnarobo\n",
    "        if end < start:\n",
    "            #reverse sequence: add +1 to reverse previous indexing correction \n",
    "            end-=1\n",
    "            start+=1\n",
    "            end2-=1\n",
    "            start2+=1\n",
    "            trna_rev=sequ[end:start]\n",
    "            ribo_rev=sequ[end2:start2]\n",
    "            trna=[]\n",
    "            rnarobo=[]\n",
    "            #reverse parts: transcribe back into the correct format!\n",
    "            for nuc in range(len(trna_rev)-1,-1,-1):\n",
    "                if trna_rev[nuc]==\"A\":\n",
    "                    trna.append(\"T\")\n",
    "                if trna_rev[nuc]==\"T\":\n",
    "                    trna.append(\"A\")\n",
    "                if trna_rev[nuc]==\"G\":\n",
    "                    trna.append(\"C\")\n",
    "                if trna_rev[nuc]==\"C\":\n",
    "                    trna.append(\"G\")\n",
    "            for nuc in range(len(ribo_rev)-1,-1,-1):\n",
    "                if ribo_rev[nuc]==\"A\":\n",
    "                    rnarobo.append(\"T\")\n",
    "                if ribo_rev[nuc]==\"T\":\n",
    "                    rnarobo.append(\"A\")\n",
    "                if ribo_rev[nuc]==\"G\":\n",
    "                    rnarobo.append(\"C\")\n",
    "                if ribo_rev[nuc]==\"C\":\n",
    "                    rnarobo.append(\"G\")\n",
    "            trna=\"\".join(trna)\n",
    "            rnarobo=\"\".join(rnarobo)\n",
    "            fff.iloc[ID,-2]=trna\n",
    "            fff.iloc[ID,-1]=rnarobo\n",
    "            \n",
    "        #output[-1][seq_beginning,seq_end]\n",
    "        #but what with the inverted ones? :/ I guess we can change them post this by inverting and changing"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 66,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>trna_nr</th>\n",
       "      <th>trna_start</th>\n",
       "      <th>trna_end</th>\n",
       "      <th>trna_type</th>\n",
       "      <th>codon</th>\n",
       "      <th>ins_start</th>\n",
       "      <th>ins_end</th>\n",
       "      <th>trna_confidence</th>\n",
       "      <th>psuedo</th>\n",
       "      <th>rnarobo_start</th>\n",
       "      <th>rnarobo_end</th>\n",
       "      <th>rnarobo_seq</th>\n",
       "      <th>trnaseq</th>\n",
       "      <th>ribozyme_control</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMEA1906455_a1_ct67271_vs1@Unclassified_virus...</td>\n",
       "      <td>10</td>\n",
       "      <td>8561</td>\n",
       "      <td>8478</td>\n",
       "      <td>Leu</td>\n",
       "      <td>TAA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>53.4</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>8477</td>\n",
       "      <td>8423</td>\n",
       "      <td>|GAATAGC|AA|ATGAGAA|ACC|T|CC|ACGT|GGT|GCTATTT|...</td>\n",
       "      <td>ACCCCTTTATCCGAACAGGTATAGGAGATGGTCTTAAACACCATTT...</td>\n",
       "      <td>GAATAGCAAATGAGAAACCTCCACGTGGTGCTATTTGGGTAATGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMEA1906455_a1_ct67271_vs1@Unclassified_virus...</td>\n",
       "      <td>1</td>\n",
       "      <td>17583</td>\n",
       "      <td>17654</td>\n",
       "      <td>Sup</td>\n",
       "      <td>CTA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17655</td>\n",
       "      <td>17710</td>\n",
       "      <td>|GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...</td>\n",
       "      <td>TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...</td>\n",
       "      <td>GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGATAACGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMEA1906581_a1_ct61757_vs1@Siphoviridae__sp__...</td>\n",
       "      <td>3</td>\n",
       "      <td>113271</td>\n",
       "      <td>113199</td>\n",
       "      <td>Met</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>53.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113198</td>\n",
       "      <td>113143</td>\n",
       "      <td>|GAGTTGC|AG|TTAGTCA|ACC|T|CC|ACGT|GGT|GCAACTT|...</td>\n",
       "      <td>CGCGGGGTATAGCAACAGGTAGCTTGTCAGTCTCATAATCTGAAGG...</td>\n",
       "      <td>GAGTTGCAGTTAGTCAACCTCCACGTGGTGCAACTTGGATTAAAAC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMEA2580118_a1_ct141914_vs1@Podoviridae__sp__...</td>\n",
       "      <td>16</td>\n",
       "      <td>33852</td>\n",
       "      <td>33767</td>\n",
       "      <td>Leu</td>\n",
       "      <td>TAA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>55.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33766</td>\n",
       "      <td>33712</td>\n",
       "      <td>|GAAACAC|AA|GATTGAA|ACC|T|CC|ACGT|GGT|GTGTTTT|...</td>\n",
       "      <td>GCCAGAATGGTGGAATTGGTAGACACGTTAGAATTAAGATCTAATG...</td>\n",
       "      <td>GAAACACAAGATTGAAACCTCCACGTGGTGTGTTTTGGGTAAAGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMEA2580118_a1_ct141914_vs1@Podoviridae__sp__...</td>\n",
       "      <td>9</td>\n",
       "      <td>41510</td>\n",
       "      <td>41431</td>\n",
       "      <td>Met</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>42.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41430</td>\n",
       "      <td>41376</td>\n",
       "      <td>|GATTCAC|AT|AGTAAGA|ACC|T|CC|ACGC|GGT|GTGAATC|...</td>\n",
       "      <td>CGTGGAGTAGAGAAGTTGGTCATCTCGCTAGGCTCATAACCTAGAA...</td>\n",
       "      <td>GATTCACATAGTAAGAACCTCCACGCGGTGTGAATCGGGTAATGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>...</th>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "      <td>...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>315</th>\n",
       "      <td>uvig_588129</td>\n",
       "      <td>1</td>\n",
       "      <td>9032</td>\n",
       "      <td>8960</td>\n",
       "      <td>Leu</td>\n",
       "      <td>CAA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>51.5</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8959</td>\n",
       "      <td>8903</td>\n",
       "      <td>|GGGACTA|AAT|ATTCTCT|ACC|T|CC|GCGC|GGT|TAGTCCT...</td>\n",
       "      <td>GCCATAGTAGAGAATTGGCATATCGACTTGACTCAAAATCAAGGGT...</td>\n",
       "      <td>GGGACTAAATATTCTCTACCTCCGCGCGGTTAGTCCTGGGCAATGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>316</th>\n",
       "      <td>uvig_590133</td>\n",
       "      <td>9</td>\n",
       "      <td>8567</td>\n",
       "      <td>8650</td>\n",
       "      <td>Leu</td>\n",
       "      <td>TAA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>61.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8651</td>\n",
       "      <td>8705</td>\n",
       "      <td>|GAATAGC|AA|ATGAGAA|ACC|T|CC|ACGT|GGT|GCTATTC|...</td>\n",
       "      <td>GGCTCAGTATCCGAATTGGTATAGGAGATGGTCTTAAACACCATTG...</td>\n",
       "      <td>GAATAGCAAATGAGAAACCTCCACGTGGTGCTATTCGGGTAACGCT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>317</th>\n",
       "      <td>uvig_590424</td>\n",
       "      <td>10</td>\n",
       "      <td>78931</td>\n",
       "      <td>78860</td>\n",
       "      <td>Glu</td>\n",
       "      <td>TTC</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>56.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>78859</td>\n",
       "      <td>78804</td>\n",
       "      <td>|GGAATAT|GAA|TCTTGGA|ACC|T|GC|ACGC|GGT|ATATTCT...</td>\n",
       "      <td>GCTCCCATGGACAAGTGGTTAAGTCCCCGCCCTTTCACGGCGGTGT...</td>\n",
       "      <td>GGAATATGAATCTTGGAACCTGCACGCGGTATATTCTGCTACTACT...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>318</th>\n",
       "      <td>uvig_594806</td>\n",
       "      <td>1</td>\n",
       "      <td>53848</td>\n",
       "      <td>53777</td>\n",
       "      <td>Sup</td>\n",
       "      <td>CTA</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>53776</td>\n",
       "      <td>53721</td>\n",
       "      <td>|GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...</td>\n",
       "      <td>TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...</td>\n",
       "      <td>GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGGTAACGC...</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>319</th>\n",
       "      <td>uvig_595877</td>\n",
       "      <td>10</td>\n",
       "      <td>8801</td>\n",
       "      <td>8872</td>\n",
       "      <td>Glu</td>\n",
       "      <td>TTC</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>56.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>8873</td>\n",
       "      <td>8928</td>\n",
       "      <td>|GGAATAT|GAA|TCTTGGA|ACC|T|GC|ACGC|GGT|ATATTCT...</td>\n",
       "      <td>GCTCCCATGGACAAGTGGTTAAGTCCCCGCCCTTTCACGGCGGTGT...</td>\n",
       "      <td>GGAATATGAATCTTGGAACCTGCACGCGGTATATTCTGCTACTACT...</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>320 rows × 15 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    ID  trna_nr  trna_start  \\\n",
       "0    SAMEA1906455_a1_ct67271_vs1@Unclassified_virus...       10        8561   \n",
       "1    SAMEA1906455_a1_ct67271_vs1@Unclassified_virus...        1       17583   \n",
       "2    SAMEA1906581_a1_ct61757_vs1@Siphoviridae__sp__...        3      113271   \n",
       "3    SAMEA2580118_a1_ct141914_vs1@Podoviridae__sp__...       16       33852   \n",
       "4    SAMEA2580118_a1_ct141914_vs1@Podoviridae__sp__...        9       41510   \n",
       "..                                                 ...      ...         ...   \n",
       "315                                        uvig_588129        1        9032   \n",
       "316                                        uvig_590133        9        8567   \n",
       "317                                        uvig_590424       10       78931   \n",
       "318                                        uvig_594806        1       53848   \n",
       "319                                        uvig_595877       10        8801   \n",
       "\n",
       "     trna_end trna_type codon  ins_start  ins_end  trna_confidence  psuedo  \\\n",
       "0        8478       Leu   TAA          0        0             53.4  pseudo   \n",
       "1       17654       Sup   CTA          0        0             60.8     NaN   \n",
       "2      113199       Met   CAT          0        0             53.9     NaN   \n",
       "3       33767       Leu   TAA          0        0             55.7     NaN   \n",
       "4       41431       Met   CAT          0        0             42.9     NaN   \n",
       "..        ...       ...   ...        ...      ...              ...     ...   \n",
       "315      8960       Leu   CAA          0        0             51.5     NaN   \n",
       "316      8650       Leu   TAA          0        0             61.1     NaN   \n",
       "317     78860       Glu   TTC          0        0             56.1     NaN   \n",
       "318     53777       Sup   CTA          0        0             60.8     NaN   \n",
       "319      8872       Glu   TTC          0        0             56.1     NaN   \n",
       "\n",
       "     rnarobo_start  rnarobo_end  \\\n",
       "0             8477         8423   \n",
       "1            17655        17710   \n",
       "2           113198       113143   \n",
       "3            33766        33712   \n",
       "4            41430        41376   \n",
       "..             ...          ...   \n",
       "315           8959         8903   \n",
       "316           8651         8705   \n",
       "317          78859        78804   \n",
       "318          53776        53721   \n",
       "319           8873         8928   \n",
       "\n",
       "                                           rnarobo_seq  \\\n",
       "0    |GAATAGC|AA|ATGAGAA|ACC|T|CC|ACGT|GGT|GCTATTT|...   \n",
       "1    |GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...   \n",
       "2    |GAGTTGC|AG|TTAGTCA|ACC|T|CC|ACGT|GGT|GCAACTT|...   \n",
       "3    |GAAACAC|AA|GATTGAA|ACC|T|CC|ACGT|GGT|GTGTTTT|...   \n",
       "4    |GATTCAC|AT|AGTAAGA|ACC|T|CC|ACGC|GGT|GTGAATC|...   \n",
       "..                                                 ...   \n",
       "315  |GGGACTA|AAT|ATTCTCT|ACC|T|CC|GCGC|GGT|TAGTCCT...   \n",
       "316  |GAATAGC|AA|ATGAGAA|ACC|T|CC|ACGT|GGT|GCTATTC|...   \n",
       "317  |GGAATAT|GAA|TCTTGGA|ACC|T|GC|ACGC|GGT|ATATTCT...   \n",
       "318  |GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...   \n",
       "319  |GGAATAT|GAA|TCTTGGA|ACC|T|GC|ACGC|GGT|ATATTCT...   \n",
       "\n",
       "                                               trnaseq  \\\n",
       "0    ACCCCTTTATCCGAACAGGTATAGGAGATGGTCTTAAACACCATTT...   \n",
       "1    TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...   \n",
       "2    CGCGGGGTATAGCAACAGGTAGCTTGTCAGTCTCATAATCTGAAGG...   \n",
       "3    GCCAGAATGGTGGAATTGGTAGACACGTTAGAATTAAGATCTAATG...   \n",
       "4    CGTGGAGTAGAGAAGTTGGTCATCTCGCTAGGCTCATAACCTAGAA...   \n",
       "..                                                 ...   \n",
       "315  GCCATAGTAGAGAATTGGCATATCGACTTGACTCAAAATCAAGGGT...   \n",
       "316  GGCTCAGTATCCGAATTGGTATAGGAGATGGTCTTAAACACCATTG...   \n",
       "317  GCTCCCATGGACAAGTGGTTAAGTCCCCGCCCTTTCACGGCGGTGT...   \n",
       "318  TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...   \n",
       "319  GCTCCCATGGACAAGTGGTTAAGTCCCCGCCCTTTCACGGCGGTGT...   \n",
       "\n",
       "                                      ribozyme_control  \n",
       "0    GAATAGCAAATGAGAAACCTCCACGTGGTGCTATTTGGGTAATGCT...  \n",
       "1    GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGATAACGC...  \n",
       "2    GAGTTGCAGTTAGTCAACCTCCACGTGGTGCAACTTGGATTAAAAC...  \n",
       "3    GAAACACAAGATTGAAACCTCCACGTGGTGTGTTTTGGGTAAAGCT...  \n",
       "4    GATTCACATAGTAAGAACCTCCACGCGGTGTGAATCGGGTAATGCT...  \n",
       "..                                                 ...  \n",
       "315  GGGACTAAATATTCTCTACCTCCGCGCGGTTAGTCCTGGGCAATGC...  \n",
       "316  GAATAGCAAATGAGAAACCTCCACGTGGTGCTATTCGGGTAACGCT...  \n",
       "317  GGAATATGAATCTTGGAACCTGCACGCGGTATATTCTGCTACTACT...  \n",
       "318  GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGGTAACGC...  \n",
       "319  GGAATATGAATCTTGGAACCTGCACGCGGTATATTCTGCTACTACT...  \n",
       "\n",
       "[320 rows x 15 columns]"
      ]
     },
     "execution_count": 66,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fff"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#still need to fix the introns!! :)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 32,
   "metadata": {},
   "outputs": [],
   "source": [
    "df1str=fff.to_csv(sep=\"\\t\",index=False)\n",
    "f= open(\"comb_trna_seq_plus_ribo_v2.csv\",\"w+\")\n",
    "f.write(df1str)\n",
    "f.close()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "#now trying to add the genomce accesion numbers to the final table..."
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 3,
   "metadata": {},
   "outputs": [],
   "source": [
    "fff=pd.read_csv(\"comb_trna_seq_plus_ribo_v2.csv\",sep=\"\\t\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 5,
   "metadata": {},
   "outputs": [],
   "source": [
    "new_genome_names=pd.read_csv(\"proj1_ribozymes_names_transcription.csv\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_name</th>\n",
       "      <th>biosample</th>\n",
       "      <th>virus_taxonomy</th>\n",
       "      <th>GenBank_Accession</th>\n",
       "      <th>contig_length</th>\n",
       "      <th>contig_topology</th>\n",
       "      <th>virus_hallmarks</th>\n",
       "      <th>bacterial_genus_crispr_host</th>\n",
       "      <th>bacterial_genus_crispr_matches</th>\n",
       "      <th>encodes_crispr_array</th>\n",
       "      <th>...</th>\n",
       "      <th>hmp_POSTERIOR_FORNIX_rpkm_mean</th>\n",
       "      <th>hmp_POSTERIOR_FORNIX_prevalence</th>\n",
       "      <th>hmp_RETROAURICULAR_CREASE_rpkm_mean</th>\n",
       "      <th>hmp_RETROAURICULAR_CREASE_prevalence</th>\n",
       "      <th>hmp_STOOL_rpkm_mean</th>\n",
       "      <th>hmp_STOOL_prevalence</th>\n",
       "      <th>hmp_SUPRAGINGIVAL_PLAQUE_rpkm_mean</th>\n",
       "      <th>hmp_SUPRAGINGIVAL_PLAQUE_prevalence</th>\n",
       "      <th>hmp_TONGUE_DORSUM_rpkm_mean</th>\n",
       "      <th>hmp_TONGUE_DORSUM_prevalence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>17910</th>\n",
       "      <td>SAMN06164915_a1_ct144655_vs1</td>\n",
       "      <td>SAMN06164915</td>\n",
       "      <td>Unclassified_virus__sp._ct88n1</td>\n",
       "      <td>BK018594</td>\n",
       "      <td>183062</td>\n",
       "      <td>linear</td>\n",
       "      <td>crass_cluster.39/PDB:3BH0_A-Helicase-ATPase-RE...</td>\n",
       "      <td>Pectobacterium</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0.001689</td>\n",
       "      <td>0.005525</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.01162</td>\n",
       "      <td>0.025532</td>\n",
       "      <td>0.000757</td>\n",
       "      <td>0.002801</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                        contig_name     biosample  \\\n",
       "17910  SAMN06164915_a1_ct144655_vs1  SAMN06164915   \n",
       "\n",
       "                       virus_taxonomy GenBank_Accession  contig_length  \\\n",
       "17910  Unclassified_virus__sp._ct88n1          BK018594         183062   \n",
       "\n",
       "      contig_topology                                    virus_hallmarks  \\\n",
       "17910          linear  crass_cluster.39/PDB:3BH0_A-Helicase-ATPase-RE...   \n",
       "\n",
       "      bacterial_genus_crispr_host  bacterial_genus_crispr_matches  \\\n",
       "17910              Pectobacterium                               1   \n",
       "\n",
       "      encodes_crispr_array  ... hmp_POSTERIOR_FORNIX_rpkm_mean  \\\n",
       "17910                   No  ...                       0.001689   \n",
       "\n",
       "      hmp_POSTERIOR_FORNIX_prevalence  hmp_RETROAURICULAR_CREASE_rpkm_mean  \\\n",
       "17910                        0.005525                                  0.0   \n",
       "\n",
       "      hmp_RETROAURICULAR_CREASE_prevalence hmp_STOOL_rpkm_mean  \\\n",
       "17910                                  0.0             0.01162   \n",
       "\n",
       "      hmp_STOOL_prevalence hmp_SUPRAGINGIVAL_PLAQUE_rpkm_mean  \\\n",
       "17910             0.025532                           0.000757   \n",
       "\n",
       "      hmp_SUPRAGINGIVAL_PLAQUE_prevalence hmp_TONGUE_DORSUM_rpkm_mean  \\\n",
       "17910                            0.002801                         0.0   \n",
       "\n",
       "      hmp_TONGUE_DORSUM_prevalence  \n",
       "17910                          0.0  \n",
       "\n",
       "[1 rows x 38 columns]"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_genome_names[new_genome_names[\"contig_name\"].str.contains(\"SAMN06164915_a1_ct144655_vs1\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 12,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_name</th>\n",
       "      <th>biosample</th>\n",
       "      <th>virus_taxonomy</th>\n",
       "      <th>GenBank_Accession</th>\n",
       "      <th>contig_length</th>\n",
       "      <th>contig_topology</th>\n",
       "      <th>virus_hallmarks</th>\n",
       "      <th>bacterial_genus_crispr_host</th>\n",
       "      <th>bacterial_genus_crispr_matches</th>\n",
       "      <th>encodes_crispr_array</th>\n",
       "      <th>...</th>\n",
       "      <th>hmp_POSTERIOR_FORNIX_rpkm_mean</th>\n",
       "      <th>hmp_POSTERIOR_FORNIX_prevalence</th>\n",
       "      <th>hmp_RETROAURICULAR_CREASE_rpkm_mean</th>\n",
       "      <th>hmp_RETROAURICULAR_CREASE_prevalence</th>\n",
       "      <th>hmp_STOOL_rpkm_mean</th>\n",
       "      <th>hmp_STOOL_prevalence</th>\n",
       "      <th>hmp_SUPRAGINGIVAL_PLAQUE_rpkm_mean</th>\n",
       "      <th>hmp_SUPRAGINGIVAL_PLAQUE_prevalence</th>\n",
       "      <th>hmp_TONGUE_DORSUM_rpkm_mean</th>\n",
       "      <th>hmp_TONGUE_DORSUM_prevalence</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>1104</th>\n",
       "      <td>SAMEA1906581_a1_ct61757_vs1</td>\n",
       "      <td>SAMEA1906581</td>\n",
       "      <td>Siphoviridae__sp._ctpon1</td>\n",
       "      <td>BK026156</td>\n",
       "      <td>128956</td>\n",
       "      <td>linear</td>\n",
       "      <td>baits_missed.22/PDB:4IDH_A-terminase-large-sub...</td>\n",
       "      <td>Campylobacter</td>\n",
       "      <td>1</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "      <td>0.0</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>1 rows × 38 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                      contig_name     biosample            virus_taxonomy  \\\n",
       "1104  SAMEA1906581_a1_ct61757_vs1  SAMEA1906581  Siphoviridae__sp._ctpon1   \n",
       "\n",
       "     GenBank_Accession  contig_length contig_topology  \\\n",
       "1104          BK026156         128956          linear   \n",
       "\n",
       "                                        virus_hallmarks  \\\n",
       "1104  baits_missed.22/PDB:4IDH_A-terminase-large-sub...   \n",
       "\n",
       "     bacterial_genus_crispr_host  bacterial_genus_crispr_matches  \\\n",
       "1104               Campylobacter                               1   \n",
       "\n",
       "     encodes_crispr_array  ... hmp_POSTERIOR_FORNIX_rpkm_mean  \\\n",
       "1104                   No  ...                            0.0   \n",
       "\n",
       "     hmp_POSTERIOR_FORNIX_prevalence  hmp_RETROAURICULAR_CREASE_rpkm_mean  \\\n",
       "1104                             0.0                                  0.0   \n",
       "\n",
       "     hmp_RETROAURICULAR_CREASE_prevalence hmp_STOOL_rpkm_mean  \\\n",
       "1104                                  0.0                 0.0   \n",
       "\n",
       "     hmp_STOOL_prevalence hmp_SUPRAGINGIVAL_PLAQUE_rpkm_mean  \\\n",
       "1104                  0.0                                0.0   \n",
       "\n",
       "     hmp_SUPRAGINGIVAL_PLAQUE_prevalence hmp_TONGUE_DORSUM_rpkm_mean  \\\n",
       "1104                                 0.0                         0.0   \n",
       "\n",
       "     hmp_TONGUE_DORSUM_prevalence  \n",
       "1104                          0.0  \n",
       "\n",
       "[1 rows x 38 columns]"
      ]
     },
     "execution_count": 12,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "new_genome_names[new_genome_names[\"contig_name\"].str.contains(\"SAMEA1906581_a1_ct61757\")]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 23,
   "metadata": {},
   "outputs": [],
   "source": [
    "#create the correct IDs for merging\n",
    "for nu in range(len(fff)):\n",
    "    fff.loc[nu,\"ID2\"]=fff[\"ID\"].str.split(\"@\")[nu][0]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "metadata": {},
   "outputs": [],
   "source": [
    "merg=new_genome_names.merge(fff, right_on=\"ID2\",left_on=\"contig_name\",how=\"right\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 39,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['GAGTTGCAGTTAGTCAACCTCCACGTGGTGCAACTTGGATTAAAACTAATGACTAA'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 39,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fff[fff[\"ID2\"]==\"SAMEA1906581_a1_ct61757_vs1\"][\"ribozyme_control\"].values"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>ID</th>\n",
       "      <th>trna_nr</th>\n",
       "      <th>trna_start</th>\n",
       "      <th>trna_end</th>\n",
       "      <th>trna_type</th>\n",
       "      <th>codon</th>\n",
       "      <th>ins_start</th>\n",
       "      <th>ins_end</th>\n",
       "      <th>trna_confidence</th>\n",
       "      <th>psuedo</th>\n",
       "      <th>rnarobo_start</th>\n",
       "      <th>rnarobo_end</th>\n",
       "      <th>rnarobo_seq</th>\n",
       "      <th>trnaseq</th>\n",
       "      <th>ribozyme_control</th>\n",
       "      <th>ID2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMEA1906581_a1_ct61757_vs1@Siphoviridae__sp__...</td>\n",
       "      <td>3</td>\n",
       "      <td>113271</td>\n",
       "      <td>113199</td>\n",
       "      <td>Met</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>53.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113198</td>\n",
       "      <td>113143</td>\n",
       "      <td>|GAGTTGC|AG|TTAGTCA|ACC|T|CC|ACGT|GGT|GCAACTT|...</td>\n",
       "      <td>CGCGGGGTATAGCAACAGGTAGCTTGTCAGTCTCATAATCTGAAGG...</td>\n",
       "      <td>GAGTTGCAGTTAGTCAACCTCCACGTGGTGCAACTTGGATTAAAAC...</td>\n",
       "      <td>SAMEA1906581_a1_ct61757_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>92</th>\n",
       "      <td>uvig_27472</td>\n",
       "      <td>3</td>\n",
       "      <td>111868</td>\n",
       "      <td>111796</td>\n",
       "      <td>Met</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>56.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>111795</td>\n",
       "      <td>111740</td>\n",
       "      <td>|GAGTTGC|AG|TTAGTCA|ACC|T|CC|ACGT|GGT|GCAACTT|...</td>\n",
       "      <td>CGCGGGGTGTAGCAACAGGTAGCTTGTCAGTCTCATAATCTGAAGG...</td>\n",
       "      <td>GAGTTGCAGTTAGTCAACCTCCACGTGGTGCAACTTGGATTAAAAC...</td>\n",
       "      <td>uvig_27472</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>121</th>\n",
       "      <td>uvig_56191</td>\n",
       "      <td>3</td>\n",
       "      <td>51827</td>\n",
       "      <td>51755</td>\n",
       "      <td>Met</td>\n",
       "      <td>CAT</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>51754</td>\n",
       "      <td>51699</td>\n",
       "      <td>|GAGTTGC|AG|TTAGTCA|ACC|T|CC|ACGT|GGT|GCAACTT|...</td>\n",
       "      <td>CGCGGGGTATAGCAATAGGTAGCTTGTCAGTCTCATAATCTGAAGG...</td>\n",
       "      <td>GAGTTGCAGTTAGTCAACCTCCACGTGGTGCAACTTGGATTAAAAC...</td>\n",
       "      <td>uvig_56191</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                                    ID  trna_nr  trna_start  \\\n",
       "2    SAMEA1906581_a1_ct61757_vs1@Siphoviridae__sp__...        3      113271   \n",
       "92                                          uvig_27472        3      111868   \n",
       "121                                         uvig_56191        3       51827   \n",
       "\n",
       "     trna_end trna_type codon  ins_start  ins_end  trna_confidence psuedo  \\\n",
       "2      113199       Met   CAT          0        0             53.9    NaN   \n",
       "92     111796       Met   CAT          0        0             56.1    NaN   \n",
       "121     51755       Met   CAT          0        0             57.1    NaN   \n",
       "\n",
       "     rnarobo_start  rnarobo_end  \\\n",
       "2           113198       113143   \n",
       "92          111795       111740   \n",
       "121          51754        51699   \n",
       "\n",
       "                                           rnarobo_seq  \\\n",
       "2    |GAGTTGC|AG|TTAGTCA|ACC|T|CC|ACGT|GGT|GCAACTT|...   \n",
       "92   |GAGTTGC|AG|TTAGTCA|ACC|T|CC|ACGT|GGT|GCAACTT|...   \n",
       "121  |GAGTTGC|AG|TTAGTCA|ACC|T|CC|ACGT|GGT|GCAACTT|...   \n",
       "\n",
       "                                               trnaseq  \\\n",
       "2    CGCGGGGTATAGCAACAGGTAGCTTGTCAGTCTCATAATCTGAAGG...   \n",
       "92   CGCGGGGTGTAGCAACAGGTAGCTTGTCAGTCTCATAATCTGAAGG...   \n",
       "121  CGCGGGGTATAGCAATAGGTAGCTTGTCAGTCTCATAATCTGAAGG...   \n",
       "\n",
       "                                      ribozyme_control  \\\n",
       "2    GAGTTGCAGTTAGTCAACCTCCACGTGGTGCAACTTGGATTAAAAC...   \n",
       "92   GAGTTGCAGTTAGTCAACCTCCACGTGGTGCAACTTGGATTAAAAC...   \n",
       "121  GAGTTGCAGTTAGTCAACCTCCACGTGGTGCAACTTGGATTAAAAC...   \n",
       "\n",
       "                             ID2  \n",
       "2    SAMEA1906581_a1_ct61757_vs1  \n",
       "92                    uvig_27472  \n",
       "121                   uvig_56191  "
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "fff[fff[\"ribozyme_control\"]==\"GAGTTGCAGTTAGTCAACCTCCACGTGGTGCAACTTGGATTAAAACTAATGACTAA\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 41,
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>contig_name</th>\n",
       "      <th>biosample</th>\n",
       "      <th>virus_taxonomy</th>\n",
       "      <th>GenBank_Accession</th>\n",
       "      <th>contig_length</th>\n",
       "      <th>contig_topology</th>\n",
       "      <th>virus_hallmarks</th>\n",
       "      <th>bacterial_genus_crispr_host</th>\n",
       "      <th>bacterial_genus_crispr_matches</th>\n",
       "      <th>encodes_crispr_array</th>\n",
       "      <th>...</th>\n",
       "      <th>ins_start</th>\n",
       "      <th>ins_end</th>\n",
       "      <th>trna_confidence</th>\n",
       "      <th>psuedo</th>\n",
       "      <th>rnarobo_start</th>\n",
       "      <th>rnarobo_end</th>\n",
       "      <th>rnarobo_seq</th>\n",
       "      <th>trnaseq</th>\n",
       "      <th>ribozyme_control</th>\n",
       "      <th>ID2</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>SAMEA1906581_a1_ct61757_vs1</td>\n",
       "      <td>SAMEA1906581</td>\n",
       "      <td>Siphoviridae__sp._ctpon1</td>\n",
       "      <td>BK026156</td>\n",
       "      <td>128956.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>baits_missed.22/PDB:4IDH_A-terminase-large-sub...</td>\n",
       "      <td>Campylobacter</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>53.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>113198</td>\n",
       "      <td>113143</td>\n",
       "      <td>|GAGTTGC|AG|TTAGTCA|ACC|T|CC|ACGT|GGT|GCAACTT|...</td>\n",
       "      <td>CGCGGGGTATAGCAACAGGTAGCTTGTCAGTCTCATAATCTGAAGG...</td>\n",
       "      <td>GAGTTGCAGTTAGTCAACCTCCACGTGGTGCAACTTGGATTAAAAC...</td>\n",
       "      <td>SAMEA1906581_a1_ct61757_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>SAMEA2580118_a1_ct141914_vs1</td>\n",
       "      <td>SAMEA2580118</td>\n",
       "      <td>Podoviridae__sp._ctiO71</td>\n",
       "      <td>BK032511</td>\n",
       "      <td>182227.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>crass_cluster.3/PDB:3EZK_B-Terminase-large-sub...</td>\n",
       "      <td>NO_HOST</td>\n",
       "      <td>0.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>55.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>33766</td>\n",
       "      <td>33712</td>\n",
       "      <td>|GAAACAC|AA|GATTGAA|ACC|T|CC|ACGT|GGT|GTGTTTT|...</td>\n",
       "      <td>GCCAGAATGGTGGAATTGGTAGACACGTTAGAATTAAGATCTAATG...</td>\n",
       "      <td>GAAACACAAGATTGAAACCTCCACGTGGTGTGTTTTGGGTAAAGCT...</td>\n",
       "      <td>SAMEA2580118_a1_ct141914_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>SAMEA2580118_a1_ct141914_vs1</td>\n",
       "      <td>SAMEA2580118</td>\n",
       "      <td>Podoviridae__sp._ctiO71</td>\n",
       "      <td>BK032511</td>\n",
       "      <td>182227.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>crass_cluster.3/PDB:3EZK_B-Terminase-large-sub...</td>\n",
       "      <td>NO_HOST</td>\n",
       "      <td>0.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>42.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>41430</td>\n",
       "      <td>41376</td>\n",
       "      <td>|GATTCAC|AT|AGTAAGA|ACC|T|CC|ACGC|GGT|GTGAATC|...</td>\n",
       "      <td>CGTGGAGTAGAGAAGTTGGTCATCTCGCTAGGCTCATAACCTAGAA...</td>\n",
       "      <td>GATTCACATAGTAAGAACCTCCACGCGGTGTGAATCGGGTAATGCT...</td>\n",
       "      <td>SAMEA2580118_a1_ct141914_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>SAMEA2737730_a1_ct2155</td>\n",
       "      <td>SAMEA2737730</td>\n",
       "      <td>Unclassified_virus__sp._ctqOj155</td>\n",
       "      <td>BK026766</td>\n",
       "      <td>189437.0</td>\n",
       "      <td>circular</td>\n",
       "      <td>crass_cluster.1/PFAM:PF17236.1-Major-Capsid-Pr...</td>\n",
       "      <td>Pectobacterium</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>59.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>142530</td>\n",
       "      <td>142584</td>\n",
       "      <td>|GAAGCAC|AA|GATTGAA|ACC|T|CC|ACGT|GGT|GTGTTTT|...</td>\n",
       "      <td>GCCCGGATGGTGGAATTGGTAGACACGTCAGATTTAAGCTCTGATG...</td>\n",
       "      <td>GAAGCACAAGATTGAAACCTCCACGTGGTGTGTTTTGGGTAAAGCT...</td>\n",
       "      <td>SAMEA2737730_a1_ct2155</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>SAMN00792023_b1_ct15_vs2</td>\n",
       "      <td>SAMN00792023</td>\n",
       "      <td>Podoviridae__sp._ctXeX15</td>\n",
       "      <td>BK049064</td>\n",
       "      <td>176478.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>some_caudo_cluster.f130/CDD:cd07386-DNA-polyme...</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>5.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>56.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>144847</td>\n",
       "      <td>144902</td>\n",
       "      <td>|GAATAGC|AAT|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...</td>\n",
       "      <td>TCTCTTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...</td>\n",
       "      <td>GAATAGCAATAGTAGAAACCTCCACGTGGTGCTATTCGGGTAACGC...</td>\n",
       "      <td>SAMN00792023_b1_ct15_vs2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>SAMN00792115_a1_ct6055</td>\n",
       "      <td>SAMN00792115</td>\n",
       "      <td>Podoviridae__sp._ct3qP055</td>\n",
       "      <td>BK051195</td>\n",
       "      <td>185487.0</td>\n",
       "      <td>circular</td>\n",
       "      <td>crass_cluster.39/PDB:3BH0_A-Helicase-ATPase-RE...</td>\n",
       "      <td>Bacteroides</td>\n",
       "      <td>3.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150820</td>\n",
       "      <td>150766</td>\n",
       "      <td>|GGTCAC|AT|AGTAGAA|ACC|T|CC|ACGT|GGT|GTGACT|GG...</td>\n",
       "      <td>TGCGGAGTGCGTAGAAATGGTATCTCACGAGGCTCATATCCTCGGG...</td>\n",
       "      <td>GGTCACATAGTAGAAACCTCCACGTGGTGTGACTGGATAACCAACC...</td>\n",
       "      <td>SAMN00792115_a1_ct6055</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>SAMN05826973_a1_ct5488</td>\n",
       "      <td>SAMN05826973</td>\n",
       "      <td>Podoviridae__sp._ctxWQ488</td>\n",
       "      <td>BK046287</td>\n",
       "      <td>184563.0</td>\n",
       "      <td>ITR</td>\n",
       "      <td>crass_cluster.3/PDB:3EZK_B-Terminase-large-sub...</td>\n",
       "      <td>Phaeodactylibacter</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>29.0</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>109622</td>\n",
       "      <td>109568</td>\n",
       "      <td>|GAATCTC|AT|AGTCGAA|ACC|T|CC|ACGT|GGT|GAGATTC|...</td>\n",
       "      <td>TGCGGGCTAGAGAAACAGTTATCTCGTCACTCTCATAAGGTGAAGA...</td>\n",
       "      <td>GAATCTCATAGTCGAAACCTCCACGTGGTGAGATTCGGGTAACGCT...</td>\n",
       "      <td>SAMN05826973_a1_ct5488</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>SAMN06164413_a1_ct151010_vs1</td>\n",
       "      <td>SAMN06164413</td>\n",
       "      <td>Podoviridae__sp._ctpL31</td>\n",
       "      <td>BK045848</td>\n",
       "      <td>119339.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>crass_cluster.16/PFAM:PF11134.7-stabilisation-...</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>56.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>15338</td>\n",
       "      <td>15393</td>\n",
       "      <td>|GAATAGC|AAT|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...</td>\n",
       "      <td>TCTCTTATGGTGTAATGGTCAGCACAGAAGACTCTAAATCTTTTAG...</td>\n",
       "      <td>GAATAGCAATAGTAGAAACCTCCACGTGGTGCTATTCGGGTAACGC...</td>\n",
       "      <td>SAMN06164413_a1_ct151010_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>SAMN06164915_a1_ct144655_vs1</td>\n",
       "      <td>SAMN06164915</td>\n",
       "      <td>Unclassified_virus__sp._ct88n1</td>\n",
       "      <td>BK018594</td>\n",
       "      <td>183062.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>crass_cluster.39/PDB:3BH0_A-Helicase-ATPase-RE...</td>\n",
       "      <td>Pectobacterium</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>66.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>69423</td>\n",
       "      <td>69477</td>\n",
       "      <td>|GAAACAC|AA|GATTGAA|ACC|T|CC|ACGT|GGT|GTGTTTT|...</td>\n",
       "      <td>GCCCGAATGGCGAAATTGGTAGACGCGTCAGATTTAAGCTCTGATG...</td>\n",
       "      <td>GAAACACAAGATTGAAACCTCCACGTGGTGTGTTTTGGGTAAAGCT...</td>\n",
       "      <td>SAMN06164915_a1_ct144655_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>SAMN09781698_a1_ct58215_vs1</td>\n",
       "      <td>SAMN09781698</td>\n",
       "      <td>Unclassified_virus__sp._ctUTx1</td>\n",
       "      <td>BK047561</td>\n",
       "      <td>83200.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>crass_cluster.3/PDB:3EZK_B-Terminase-large-sub...</td>\n",
       "      <td>Eubacterium</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>666</td>\n",
       "      <td>721</td>\n",
       "      <td>|GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...</td>\n",
       "      <td>TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...</td>\n",
       "      <td>GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGGTAACGC...</td>\n",
       "      <td>SAMN09781698_a1_ct58215_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>10</th>\n",
       "      <td>SAMN09781742_a1_ct130054_vs1</td>\n",
       "      <td>SAMN09781742</td>\n",
       "      <td>Unclassified_virus__sp._ctKf81</td>\n",
       "      <td>BK017706</td>\n",
       "      <td>46728.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>crass_cluster.3/PDB:3EZK_B-Terminase-large-sub...</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>2.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>56.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13154</td>\n",
       "      <td>13209</td>\n",
       "      <td>|GAATAGC|AAT|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTT...</td>\n",
       "      <td>TCTCTTATGGTGTAATGGTCAGCACAGAAGACTCTAAATCTTTTAG...</td>\n",
       "      <td>GAATAGCAATAGTAGAAACCTCCACGTGGTGCTATTTGGGTAATGC...</td>\n",
       "      <td>SAMN09781742_a1_ct130054_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>11</th>\n",
       "      <td>SAMN09781762_a1_ct159411_vs1</td>\n",
       "      <td>SAMN09781762</td>\n",
       "      <td>Myoviridae__sp._ct5WI1</td>\n",
       "      <td>BK054677</td>\n",
       "      <td>19877.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>baits_missed.250/PDB:5IV5_HI-Baseplate-wedge-p...</td>\n",
       "      <td>NO_HOST</td>\n",
       "      <td>0.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>41.6</td>\n",
       "      <td>NaN</td>\n",
       "      <td>18762</td>\n",
       "      <td>18710</td>\n",
       "      <td>|GGTGCT|AA|AAACAAA|ACC|T|CC|TCGT|GGT|AGCACT|GG...</td>\n",
       "      <td>TGAGGGGTAGCATAGCGGCAAATGCGCCGGTCTCTAAAACCGGATA...</td>\n",
       "      <td>GGTGCTAAAAACAAAACCTCCTCGTGGTAGCACTGGGTAATGCTAA...</td>\n",
       "      <td>SAMN09781762_a1_ct159411_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>12</th>\n",
       "      <td>SAMN09781783_a1_ct211607_vs1</td>\n",
       "      <td>SAMN09781783</td>\n",
       "      <td>Unclassified_virus__sp._ctqks1</td>\n",
       "      <td>BK048594</td>\n",
       "      <td>54417.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>crass_cluster.3/PDB:3EZK_B-Terminase-large-sub...</td>\n",
       "      <td>NO_HOST</td>\n",
       "      <td>0.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>642</td>\n",
       "      <td>697</td>\n",
       "      <td>|GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...</td>\n",
       "      <td>TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...</td>\n",
       "      <td>GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGGTAACGC...</td>\n",
       "      <td>SAMN09781783_a1_ct211607_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>13</th>\n",
       "      <td>SAMN10080920_a1_ct29203_vs1</td>\n",
       "      <td>SAMN10080920</td>\n",
       "      <td>Podoviridae__sp._ctzd31</td>\n",
       "      <td>BK035234</td>\n",
       "      <td>109620.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>some_caudo_cluster.f130/CDD:cd07386-DNA-polyme...</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>2.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>44.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>90998</td>\n",
       "      <td>90944</td>\n",
       "      <td>|GATTCAC|AT|AGTAAGA|ACC|T|CC|ACGT|GGT|GTGAGTT|...</td>\n",
       "      <td>GCGGATGTGGTGTAATGGTAACATACTAGGCTCATAACCTAGAGTA...</td>\n",
       "      <td>GATTCACATAGTAAGAACCTCCACGTGGTGTGAGTTGGGTAATGCT...</td>\n",
       "      <td>SAMN10080920_a1_ct29203_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>14</th>\n",
       "      <td>SAMN11252288_a1_ct6785</td>\n",
       "      <td>SAMN11252288</td>\n",
       "      <td>Unclassified_virus__sp._ctpci785</td>\n",
       "      <td>BK016703</td>\n",
       "      <td>170455.0</td>\n",
       "      <td>ITR</td>\n",
       "      <td>crass_cluster.1/PFAM:PF17236.1-Major-Capsid-Pr...</td>\n",
       "      <td>Prevotella</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>46.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>25727</td>\n",
       "      <td>25782</td>\n",
       "      <td>|GAAGTGT|GT|GTAAGCA|ACC|T|CC|ACGC|GGT|ACACTTC|...</td>\n",
       "      <td>CTAGGTGTCGGTCAATAGGTAGGCTCCTGCATTTGGGATGCAGATG...</td>\n",
       "      <td>GAAGTGTGTGTAAGCAACCTCCACGCGGTACACTTCGGGTAACTCT...</td>\n",
       "      <td>SAMN11252288_a1_ct6785</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>15</th>\n",
       "      <td>SAMN11893770_a1_ct7896_vs1</td>\n",
       "      <td>SAMN11893770</td>\n",
       "      <td>Siphoviridae__sp._ctgcT1</td>\n",
       "      <td>BK014771</td>\n",
       "      <td>196584.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>baits_missed.9/PFAM:PF10145.8-minor-tail-prote...</td>\n",
       "      <td>Listeria</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>35.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>32654</td>\n",
       "      <td>32708</td>\n",
       "      <td>|GGGAAC|TAT|CCACAAT|ACC|T|TC|ACGC|GGT|GTTCCT|G...</td>\n",
       "      <td>AGGCGTCTGGTGTAACGGTCGCATTCGGATCTCCAAAATCTTGAGG...</td>\n",
       "      <td>GGGAACTATCCACAATACCTTCACGCGGTGTTCCTGATTTCGACTA...</td>\n",
       "      <td>SAMN11893770_a1_ct7896_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>16</th>\n",
       "      <td>SAMN11893770_a1_ct7896_vs1</td>\n",
       "      <td>SAMN11893770</td>\n",
       "      <td>Siphoviridae__sp._ctgcT1</td>\n",
       "      <td>BK014771</td>\n",
       "      <td>196584.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>baits_missed.9/PFAM:PF10145.8-minor-tail-prote...</td>\n",
       "      <td>Listeria</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>42741</td>\n",
       "      <td>42793</td>\n",
       "      <td>|GGTGCT|AA|AAACAAA|ACC|T|CC|TCGC|GGT|AGCACT|GG...</td>\n",
       "      <td>TGCCCCGTAGTGTAACGGTTAACACGGCGGTCTCTAAAACCGTTTT...</td>\n",
       "      <td>GGTGCTAAAAACAAAACCTCCTCGCGGTAGCACTGGGTAATGCTAA...</td>\n",
       "      <td>SAMN11893770_a1_ct7896_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>17</th>\n",
       "      <td>SRS015484_a1_ct4036</td>\n",
       "      <td>SRS015484</td>\n",
       "      <td>Siphoviridae__sp._ctKwe036</td>\n",
       "      <td>BK049013</td>\n",
       "      <td>99134.0</td>\n",
       "      <td>circular</td>\n",
       "      <td>baits_missed.45/UniProtKB:E0YJ23_9CAUD-Putativ...</td>\n",
       "      <td>NO_HOST</td>\n",
       "      <td>0.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>50.7</td>\n",
       "      <td>NaN</td>\n",
       "      <td>36442</td>\n",
       "      <td>36498</td>\n",
       "      <td>|GAGACAC|AAG|ATTATGT|ACT|T|CC|ACGC|AGT|GTGTCTT...</td>\n",
       "      <td>TGCGGAGTATTAGCAGTGGTAGCTGGCTAGGTTCATACCCTAGAGG...</td>\n",
       "      <td>GAGACACAAGATTATGTACTTCCACGCAGTGTGTCTTGGGTAATTC...</td>\n",
       "      <td>SRS015484_a1_ct4036</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>18</th>\n",
       "      <td>SRS017375_a1_ct1878</td>\n",
       "      <td>SRS017375</td>\n",
       "      <td>Unclassified_virus__sp._ctRPv878</td>\n",
       "      <td>BK056179</td>\n",
       "      <td>178799.0</td>\n",
       "      <td>circular</td>\n",
       "      <td>crass_cluster.1/PFAM:PF17236.1-Major-Capsid-Pr...</td>\n",
       "      <td>Eubacterium</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>53.4</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>8380</td>\n",
       "      <td>8434</td>\n",
       "      <td>|GAATAGC|AA|ATGAGAA|ACC|T|CC|ACGT|GGT|GCTATTT|...</td>\n",
       "      <td>ACCCCTTTATCCGAACAGGTATAGGAGATGGTCTTAAACACCATTT...</td>\n",
       "      <td>GAATAGCAAATGAGAAACCTCCACGTGGTGCTATTTGGGTAACGCT...</td>\n",
       "      <td>SRS017375_a1_ct1878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>19</th>\n",
       "      <td>SRS017375_a1_ct1878</td>\n",
       "      <td>SRS017375</td>\n",
       "      <td>Unclassified_virus__sp._ctRPv878</td>\n",
       "      <td>BK056179</td>\n",
       "      <td>178799.0</td>\n",
       "      <td>circular</td>\n",
       "      <td>crass_cluster.1/PFAM:PF17236.1-Major-Capsid-Pr...</td>\n",
       "      <td>Eubacterium</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>57.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>176114</td>\n",
       "      <td>176059</td>\n",
       "      <td>|GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...</td>\n",
       "      <td>TCTCCTATGGTGTAATGGTTTAGCACAGAAGACTCTAAATCTTTTA...</td>\n",
       "      <td>GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGATAACGC...</td>\n",
       "      <td>SRS017375_a1_ct1878</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>20</th>\n",
       "      <td>SRS076804_a1_ct18023_vs1</td>\n",
       "      <td>SRS076804</td>\n",
       "      <td>Siphoviridae__sp._ctLda1</td>\n",
       "      <td>BK046446</td>\n",
       "      <td>123303.0</td>\n",
       "      <td>linear</td>\n",
       "      <td>baits_missed.46/PDB:2FKK_A-Baseplate|baits_mis...</td>\n",
       "      <td>Eubacterium</td>\n",
       "      <td>1.0</td>\n",
       "      <td>No</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>54.4</td>\n",
       "      <td>NaN</td>\n",
       "      <td>117932</td>\n",
       "      <td>117987</td>\n",
       "      <td>|GGGAAAC|TAT|ATTCTCA|ACC|T|GC|TCGT|GGT|GTTTCCT...</td>\n",
       "      <td>GCACCTATAATTCAATTGGCAGAATAGCTGACTTGTAATCAGCCTA...</td>\n",
       "      <td>GGGAAACTATATTCTCAACCTGCTCGTGGTGTTTCCTGCTTTTACT...</td>\n",
       "      <td>SRS076804_a1_ct18023_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>21</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>53.4</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>8477</td>\n",
       "      <td>8423</td>\n",
       "      <td>|GAATAGC|AA|ATGAGAA|ACC|T|CC|ACGT|GGT|GCTATTT|...</td>\n",
       "      <td>ACCCCTTTATCCGAACAGGTATAGGAGATGGTCTTAAACACCATTT...</td>\n",
       "      <td>GAATAGCAAATGAGAAACCTCCACGTGGTGCTATTTGGGTAATGCT...</td>\n",
       "      <td>SAMEA1906455_a1_ct67271_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>22</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60.8</td>\n",
       "      <td>NaN</td>\n",
       "      <td>17655</td>\n",
       "      <td>17710</td>\n",
       "      <td>|GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...</td>\n",
       "      <td>TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...</td>\n",
       "      <td>GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGATAACGC...</td>\n",
       "      <td>SAMEA1906455_a1_ct67271_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>23</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>56.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>6382</td>\n",
       "      <td>6437</td>\n",
       "      <td>|GAATAGC|AAT|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...</td>\n",
       "      <td>TCTCTTATGGTGTAATGGTCAGCACAGAAGACTCTAAATCTTTTAG...</td>\n",
       "      <td>GAATAGCAATAGTAGAAACCTCCACGTGGTGCTATTCGGGTAATGC...</td>\n",
       "      <td>SAMEA2580244_a1_ct185502_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>24</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>51.9</td>\n",
       "      <td>NaN</td>\n",
       "      <td>54486</td>\n",
       "      <td>54541</td>\n",
       "      <td>|GGTCAC|TAA|AGTAGAT|ACT|T|CC|ACGT|AGT|GTGACT|G...</td>\n",
       "      <td>AGCGGGATGGCGCAAAGGTAGCGCGTAGCTTTCACGAGGCTGAGGT...</td>\n",
       "      <td>GGTCACTAAAGTAGATACTTCCACGTAGTGTGACTGGATTAAAACT...</td>\n",
       "      <td>SAMEA2737658_a1_ct63307_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>25</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>42.1</td>\n",
       "      <td>NaN</td>\n",
       "      <td>70149</td>\n",
       "      <td>70096</td>\n",
       "      <td>|GACTCAC|A|AGTAAGA|ACC|T|CC|ACGT|GGT|GTGAGTT|G...</td>\n",
       "      <td>CGTGGAGTAGAGAAGTGGTCATCTCGCTAGGCTCATAACCTAGAAA...</td>\n",
       "      <td>GACTCACAAGTAAGAACCTCCACGTGGTGTGAGTTGGATAACGCTA...</td>\n",
       "      <td>SAMEA2737747_a1_ct8487</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>26</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>6488</td>\n",
       "      <td>6710</td>\n",
       "      <td>36.6</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>6746</td>\n",
       "      <td>6801</td>\n",
       "      <td>|GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...</td>\n",
       "      <td>TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTTAAAATAAGGA...</td>\n",
       "      <td>GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGATAACGC...</td>\n",
       "      <td>SAMEA2737752_a1_ct159360_vs1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>27</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>47.2</td>\n",
       "      <td>NaN</td>\n",
       "      <td>150723</td>\n",
       "      <td>150669</td>\n",
       "      <td>|GGTCAC|AT|AGTAGAA|ACC|T|CC|ACGT|GGT|GTGACT|GG...</td>\n",
       "      <td>TGCGGAGTGCGTAGAAATGGTATCTCACGAGGCTCATATCCTCGGG...</td>\n",
       "      <td>GGTCACATAGTAGAAACCTCCACGTGGTGTGACTGGATAACCAACC...</td>\n",
       "      <td>SAMEA2737813_a1_ct27500</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>28</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>23.8</td>\n",
       "      <td>pseudo</td>\n",
       "      <td>160097</td>\n",
       "      <td>160043</td>\n",
       "      <td>|GAATCTC|AT|AGTCGAA|ACC|T|CC|ACGT|GGT|GAGATTC|...</td>\n",
       "      <td>GCGGAGGAATGAAACGGATATATAAATCATACTAGTCTCATAAGCT...</td>\n",
       "      <td>GAATCTCATAGTCGAAACCTCCACGTGGTGAGATTCGGGTAACGCT...</td>\n",
       "      <td>SAMN00791959_a1_ct12730</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>29</th>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>...</td>\n",
       "      <td>0</td>\n",
       "      <td>0</td>\n",
       "      <td>60.0</td>\n",
       "      <td>NaN</td>\n",
       "      <td>13263</td>\n",
       "      <td>13318</td>\n",
       "      <td>|GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTT...</td>\n",
       "      <td>TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...</td>\n",
       "      <td>GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTTGGGTAACGC...</td>\n",
       "      <td>SAMN00993245_a1_ct62305_vs1</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "<p>30 rows × 54 columns</p>\n",
       "</div>"
      ],
      "text/plain": [
       "                     contig_name     biosample  \\\n",
       "0    SAMEA1906581_a1_ct61757_vs1  SAMEA1906581   \n",
       "1   SAMEA2580118_a1_ct141914_vs1  SAMEA2580118   \n",
       "2   SAMEA2580118_a1_ct141914_vs1  SAMEA2580118   \n",
       "3         SAMEA2737730_a1_ct2155  SAMEA2737730   \n",
       "4       SAMN00792023_b1_ct15_vs2  SAMN00792023   \n",
       "5         SAMN00792115_a1_ct6055  SAMN00792115   \n",
       "6         SAMN05826973_a1_ct5488  SAMN05826973   \n",
       "7   SAMN06164413_a1_ct151010_vs1  SAMN06164413   \n",
       "8   SAMN06164915_a1_ct144655_vs1  SAMN06164915   \n",
       "9    SAMN09781698_a1_ct58215_vs1  SAMN09781698   \n",
       "10  SAMN09781742_a1_ct130054_vs1  SAMN09781742   \n",
       "11  SAMN09781762_a1_ct159411_vs1  SAMN09781762   \n",
       "12  SAMN09781783_a1_ct211607_vs1  SAMN09781783   \n",
       "13   SAMN10080920_a1_ct29203_vs1  SAMN10080920   \n",
       "14        SAMN11252288_a1_ct6785  SAMN11252288   \n",
       "15    SAMN11893770_a1_ct7896_vs1  SAMN11893770   \n",
       "16    SAMN11893770_a1_ct7896_vs1  SAMN11893770   \n",
       "17           SRS015484_a1_ct4036     SRS015484   \n",
       "18           SRS017375_a1_ct1878     SRS017375   \n",
       "19           SRS017375_a1_ct1878     SRS017375   \n",
       "20      SRS076804_a1_ct18023_vs1     SRS076804   \n",
       "21                           NaN           NaN   \n",
       "22                           NaN           NaN   \n",
       "23                           NaN           NaN   \n",
       "24                           NaN           NaN   \n",
       "25                           NaN           NaN   \n",
       "26                           NaN           NaN   \n",
       "27                           NaN           NaN   \n",
       "28                           NaN           NaN   \n",
       "29                           NaN           NaN   \n",
       "\n",
       "                      virus_taxonomy GenBank_Accession  contig_length  \\\n",
       "0           Siphoviridae__sp._ctpon1          BK026156       128956.0   \n",
       "1            Podoviridae__sp._ctiO71          BK032511       182227.0   \n",
       "2            Podoviridae__sp._ctiO71          BK032511       182227.0   \n",
       "3   Unclassified_virus__sp._ctqOj155          BK026766       189437.0   \n",
       "4           Podoviridae__sp._ctXeX15          BK049064       176478.0   \n",
       "5          Podoviridae__sp._ct3qP055          BK051195       185487.0   \n",
       "6          Podoviridae__sp._ctxWQ488          BK046287       184563.0   \n",
       "7            Podoviridae__sp._ctpL31          BK045848       119339.0   \n",
       "8     Unclassified_virus__sp._ct88n1          BK018594       183062.0   \n",
       "9     Unclassified_virus__sp._ctUTx1          BK047561        83200.0   \n",
       "10    Unclassified_virus__sp._ctKf81          BK017706        46728.0   \n",
       "11            Myoviridae__sp._ct5WI1          BK054677        19877.0   \n",
       "12    Unclassified_virus__sp._ctqks1          BK048594        54417.0   \n",
       "13           Podoviridae__sp._ctzd31          BK035234       109620.0   \n",
       "14  Unclassified_virus__sp._ctpci785          BK016703       170455.0   \n",
       "15          Siphoviridae__sp._ctgcT1          BK014771       196584.0   \n",
       "16          Siphoviridae__sp._ctgcT1          BK014771       196584.0   \n",
       "17        Siphoviridae__sp._ctKwe036          BK049013        99134.0   \n",
       "18  Unclassified_virus__sp._ctRPv878          BK056179       178799.0   \n",
       "19  Unclassified_virus__sp._ctRPv878          BK056179       178799.0   \n",
       "20          Siphoviridae__sp._ctLda1          BK046446       123303.0   \n",
       "21                               NaN               NaN            NaN   \n",
       "22                               NaN               NaN            NaN   \n",
       "23                               NaN               NaN            NaN   \n",
       "24                               NaN               NaN            NaN   \n",
       "25                               NaN               NaN            NaN   \n",
       "26                               NaN               NaN            NaN   \n",
       "27                               NaN               NaN            NaN   \n",
       "28                               NaN               NaN            NaN   \n",
       "29                               NaN               NaN            NaN   \n",
       "\n",
       "   contig_topology                                    virus_hallmarks  \\\n",
       "0           linear  baits_missed.22/PDB:4IDH_A-terminase-large-sub...   \n",
       "1           linear  crass_cluster.3/PDB:3EZK_B-Terminase-large-sub...   \n",
       "2           linear  crass_cluster.3/PDB:3EZK_B-Terminase-large-sub...   \n",
       "3         circular  crass_cluster.1/PFAM:PF17236.1-Major-Capsid-Pr...   \n",
       "4           linear  some_caudo_cluster.f130/CDD:cd07386-DNA-polyme...   \n",
       "5         circular  crass_cluster.39/PDB:3BH0_A-Helicase-ATPase-RE...   \n",
       "6              ITR  crass_cluster.3/PDB:3EZK_B-Terminase-large-sub...   \n",
       "7           linear  crass_cluster.16/PFAM:PF11134.7-stabilisation-...   \n",
       "8           linear  crass_cluster.39/PDB:3BH0_A-Helicase-ATPase-RE...   \n",
       "9           linear  crass_cluster.3/PDB:3EZK_B-Terminase-large-sub...   \n",
       "10          linear  crass_cluster.3/PDB:3EZK_B-Terminase-large-sub...   \n",
       "11          linear  baits_missed.250/PDB:5IV5_HI-Baseplate-wedge-p...   \n",
       "12          linear  crass_cluster.3/PDB:3EZK_B-Terminase-large-sub...   \n",
       "13          linear  some_caudo_cluster.f130/CDD:cd07386-DNA-polyme...   \n",
       "14             ITR  crass_cluster.1/PFAM:PF17236.1-Major-Capsid-Pr...   \n",
       "15          linear  baits_missed.9/PFAM:PF10145.8-minor-tail-prote...   \n",
       "16          linear  baits_missed.9/PFAM:PF10145.8-minor-tail-prote...   \n",
       "17        circular  baits_missed.45/UniProtKB:E0YJ23_9CAUD-Putativ...   \n",
       "18        circular  crass_cluster.1/PFAM:PF17236.1-Major-Capsid-Pr...   \n",
       "19        circular  crass_cluster.1/PFAM:PF17236.1-Major-Capsid-Pr...   \n",
       "20          linear  baits_missed.46/PDB:2FKK_A-Baseplate|baits_mis...   \n",
       "21             NaN                                                NaN   \n",
       "22             NaN                                                NaN   \n",
       "23             NaN                                                NaN   \n",
       "24             NaN                                                NaN   \n",
       "25             NaN                                                NaN   \n",
       "26             NaN                                                NaN   \n",
       "27             NaN                                                NaN   \n",
       "28             NaN                                                NaN   \n",
       "29             NaN                                                NaN   \n",
       "\n",
       "   bacterial_genus_crispr_host  bacterial_genus_crispr_matches  \\\n",
       "0                Campylobacter                             1.0   \n",
       "1                      NO_HOST                             0.0   \n",
       "2                      NO_HOST                             0.0   \n",
       "3               Pectobacterium                             1.0   \n",
       "4                   Prevotella                             5.0   \n",
       "5                  Bacteroides                             3.0   \n",
       "6           Phaeodactylibacter                             1.0   \n",
       "7                   Prevotella                             1.0   \n",
       "8               Pectobacterium                             1.0   \n",
       "9                  Eubacterium                             1.0   \n",
       "10                  Prevotella                             2.0   \n",
       "11                     NO_HOST                             0.0   \n",
       "12                     NO_HOST                             0.0   \n",
       "13                  Prevotella                             2.0   \n",
       "14                  Prevotella                             1.0   \n",
       "15                    Listeria                             1.0   \n",
       "16                    Listeria                             1.0   \n",
       "17                     NO_HOST                             0.0   \n",
       "18                 Eubacterium                             1.0   \n",
       "19                 Eubacterium                             1.0   \n",
       "20                 Eubacterium                             1.0   \n",
       "21                         NaN                             NaN   \n",
       "22                         NaN                             NaN   \n",
       "23                         NaN                             NaN   \n",
       "24                         NaN                             NaN   \n",
       "25                         NaN                             NaN   \n",
       "26                         NaN                             NaN   \n",
       "27                         NaN                             NaN   \n",
       "28                         NaN                             NaN   \n",
       "29                         NaN                             NaN   \n",
       "\n",
       "   encodes_crispr_array  ... ins_start ins_end  trna_confidence  psuedo  \\\n",
       "0                    No  ...         0       0             53.9     NaN   \n",
       "1                    No  ...         0       0             55.7     NaN   \n",
       "2                    No  ...         0       0             42.9     NaN   \n",
       "3                    No  ...         0       0             59.9     NaN   \n",
       "4                    No  ...         0       0             56.2     NaN   \n",
       "5                    No  ...         0       0             47.2     NaN   \n",
       "6                    No  ...         0       0             29.0  pseudo   \n",
       "7                    No  ...         0       0             56.2     NaN   \n",
       "8                    No  ...         0       0             66.7     NaN   \n",
       "9                    No  ...         0       0             60.8     NaN   \n",
       "10                   No  ...         0       0             56.2     NaN   \n",
       "11                   No  ...         0       0             41.6     NaN   \n",
       "12                   No  ...         0       0             60.8     NaN   \n",
       "13                   No  ...         0       0             44.7     NaN   \n",
       "14                   No  ...         0       0             46.1     NaN   \n",
       "15                   No  ...         0       0             35.8     NaN   \n",
       "16                   No  ...         0       0             54.8     NaN   \n",
       "17                   No  ...         0       0             50.7     NaN   \n",
       "18                   No  ...         0       0             53.4  pseudo   \n",
       "19                   No  ...         0       0             57.8     NaN   \n",
       "20                   No  ...         0       0             54.4     NaN   \n",
       "21                  NaN  ...         0       0             53.4  pseudo   \n",
       "22                  NaN  ...         0       0             60.8     NaN   \n",
       "23                  NaN  ...         0       0             56.2     NaN   \n",
       "24                  NaN  ...         0       0             51.9     NaN   \n",
       "25                  NaN  ...         0       0             42.1     NaN   \n",
       "26                  NaN  ...      6488    6710             36.6  pseudo   \n",
       "27                  NaN  ...         0       0             47.2     NaN   \n",
       "28                  NaN  ...         0       0             23.8  pseudo   \n",
       "29                  NaN  ...         0       0             60.0     NaN   \n",
       "\n",
       "   rnarobo_start rnarobo_end  \\\n",
       "0         113198      113143   \n",
       "1          33766       33712   \n",
       "2          41430       41376   \n",
       "3         142530      142584   \n",
       "4         144847      144902   \n",
       "5         150820      150766   \n",
       "6         109622      109568   \n",
       "7          15338       15393   \n",
       "8          69423       69477   \n",
       "9            666         721   \n",
       "10         13154       13209   \n",
       "11         18762       18710   \n",
       "12           642         697   \n",
       "13         90998       90944   \n",
       "14         25727       25782   \n",
       "15         32654       32708   \n",
       "16         42741       42793   \n",
       "17         36442       36498   \n",
       "18          8380        8434   \n",
       "19        176114      176059   \n",
       "20        117932      117987   \n",
       "21          8477        8423   \n",
       "22         17655       17710   \n",
       "23          6382        6437   \n",
       "24         54486       54541   \n",
       "25         70149       70096   \n",
       "26          6746        6801   \n",
       "27        150723      150669   \n",
       "28        160097      160043   \n",
       "29         13263       13318   \n",
       "\n",
       "                                          rnarobo_seq  \\\n",
       "0   |GAGTTGC|AG|TTAGTCA|ACC|T|CC|ACGT|GGT|GCAACTT|...   \n",
       "1   |GAAACAC|AA|GATTGAA|ACC|T|CC|ACGT|GGT|GTGTTTT|...   \n",
       "2   |GATTCAC|AT|AGTAAGA|ACC|T|CC|ACGC|GGT|GTGAATC|...   \n",
       "3   |GAAGCAC|AA|GATTGAA|ACC|T|CC|ACGT|GGT|GTGTTTT|...   \n",
       "4   |GAATAGC|AAT|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...   \n",
       "5   |GGTCAC|AT|AGTAGAA|ACC|T|CC|ACGT|GGT|GTGACT|GG...   \n",
       "6   |GAATCTC|AT|AGTCGAA|ACC|T|CC|ACGT|GGT|GAGATTC|...   \n",
       "7   |GAATAGC|AAT|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...   \n",
       "8   |GAAACAC|AA|GATTGAA|ACC|T|CC|ACGT|GGT|GTGTTTT|...   \n",
       "9   |GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...   \n",
       "10  |GAATAGC|AAT|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTT...   \n",
       "11  |GGTGCT|AA|AAACAAA|ACC|T|CC|TCGT|GGT|AGCACT|GG...   \n",
       "12  |GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...   \n",
       "13  |GATTCAC|AT|AGTAAGA|ACC|T|CC|ACGT|GGT|GTGAGTT|...   \n",
       "14  |GAAGTGT|GT|GTAAGCA|ACC|T|CC|ACGC|GGT|ACACTTC|...   \n",
       "15  |GGGAAC|TAT|CCACAAT|ACC|T|TC|ACGC|GGT|GTTCCT|G...   \n",
       "16  |GGTGCT|AA|AAACAAA|ACC|T|CC|TCGC|GGT|AGCACT|GG...   \n",
       "17  |GAGACAC|AAG|ATTATGT|ACT|T|CC|ACGC|AGT|GTGTCTT...   \n",
       "18  |GAATAGC|AA|ATGAGAA|ACC|T|CC|ACGT|GGT|GCTATTT|...   \n",
       "19  |GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...   \n",
       "20  |GGGAAAC|TAT|ATTCTCA|ACC|T|GC|TCGT|GGT|GTTTCCT...   \n",
       "21  |GAATAGC|AA|ATGAGAA|ACC|T|CC|ACGT|GGT|GCTATTT|...   \n",
       "22  |GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...   \n",
       "23  |GAATAGC|AAT|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...   \n",
       "24  |GGTCAC|TAA|AGTAGAT|ACT|T|CC|ACGT|AGT|GTGACT|G...   \n",
       "25  |GACTCAC|A|AGTAAGA|ACC|T|CC|ACGT|GGT|GTGAGTT|G...   \n",
       "26  |GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTC...   \n",
       "27  |GGTCAC|AT|AGTAGAA|ACC|T|CC|ACGT|GGT|GTGACT|GG...   \n",
       "28  |GAATCTC|AT|AGTCGAA|ACC|T|CC|ACGT|GGT|GAGATTC|...   \n",
       "29  |GAATAGC|GTA|AGTAGAA|ACC|T|CC|ACGT|GGT|GCTATTT...   \n",
       "\n",
       "                                              trnaseq  \\\n",
       "0   CGCGGGGTATAGCAACAGGTAGCTTGTCAGTCTCATAATCTGAAGG...   \n",
       "1   GCCAGAATGGTGGAATTGGTAGACACGTTAGAATTAAGATCTAATG...   \n",
       "2   CGTGGAGTAGAGAAGTTGGTCATCTCGCTAGGCTCATAACCTAGAA...   \n",
       "3   GCCCGGATGGTGGAATTGGTAGACACGTCAGATTTAAGCTCTGATG...   \n",
       "4   TCTCTTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...   \n",
       "5   TGCGGAGTGCGTAGAAATGGTATCTCACGAGGCTCATATCCTCGGG...   \n",
       "6   TGCGGGCTAGAGAAACAGTTATCTCGTCACTCTCATAAGGTGAAGA...   \n",
       "7   TCTCTTATGGTGTAATGGTCAGCACAGAAGACTCTAAATCTTTTAG...   \n",
       "8   GCCCGAATGGCGAAATTGGTAGACGCGTCAGATTTAAGCTCTGATG...   \n",
       "9   TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...   \n",
       "10  TCTCTTATGGTGTAATGGTCAGCACAGAAGACTCTAAATCTTTTAG...   \n",
       "11  TGAGGGGTAGCATAGCGGCAAATGCGCCGGTCTCTAAAACCGGATA...   \n",
       "12  TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...   \n",
       "13  GCGGATGTGGTGTAATGGTAACATACTAGGCTCATAACCTAGAGTA...   \n",
       "14  CTAGGTGTCGGTCAATAGGTAGGCTCCTGCATTTGGGATGCAGATG...   \n",
       "15  AGGCGTCTGGTGTAACGGTCGCATTCGGATCTCCAAAATCTTGAGG...   \n",
       "16  TGCCCCGTAGTGTAACGGTTAACACGGCGGTCTCTAAAACCGTTTT...   \n",
       "17  TGCGGAGTATTAGCAGTGGTAGCTGGCTAGGTTCATACCCTAGAGG...   \n",
       "18  ACCCCTTTATCCGAACAGGTATAGGAGATGGTCTTAAACACCATTT...   \n",
       "19  TCTCCTATGGTGTAATGGTTTAGCACAGAAGACTCTAAATCTTTTA...   \n",
       "20  GCACCTATAATTCAATTGGCAGAATAGCTGACTTGTAATCAGCCTA...   \n",
       "21  ACCCCTTTATCCGAACAGGTATAGGAGATGGTCTTAAACACCATTT...   \n",
       "22  TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...   \n",
       "23  TCTCTTATGGTGTAATGGTCAGCACAGAAGACTCTAAATCTTTTAG...   \n",
       "24  AGCGGGATGGCGCAAAGGTAGCGCGTAGCTTTCACGAGGCTGAGGT...   \n",
       "25  CGTGGAGTAGAGAAGTGGTCATCTCGCTAGGCTCATAACCTAGAAA...   \n",
       "26  TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTTAAAATAAGGA...   \n",
       "27  TGCGGAGTGCGTAGAAATGGTATCTCACGAGGCTCATATCCTCGGG...   \n",
       "28  GCGGAGGAATGAAACGGATATATAAATCATACTAGTCTCATAAGCT...   \n",
       "29  TCTCCTATGGTGTAATGGTTAGCACAGAAGACTCTAAATCTTTTAG...   \n",
       "\n",
       "                                     ribozyme_control  \\\n",
       "0   GAGTTGCAGTTAGTCAACCTCCACGTGGTGCAACTTGGATTAAAAC...   \n",
       "1   GAAACACAAGATTGAAACCTCCACGTGGTGTGTTTTGGGTAAAGCT...   \n",
       "2   GATTCACATAGTAAGAACCTCCACGCGGTGTGAATCGGGTAATGCT...   \n",
       "3   GAAGCACAAGATTGAAACCTCCACGTGGTGTGTTTTGGGTAAAGCT...   \n",
       "4   GAATAGCAATAGTAGAAACCTCCACGTGGTGCTATTCGGGTAACGC...   \n",
       "5   GGTCACATAGTAGAAACCTCCACGTGGTGTGACTGGATAACCAACC...   \n",
       "6   GAATCTCATAGTCGAAACCTCCACGTGGTGAGATTCGGGTAACGCT...   \n",
       "7   GAATAGCAATAGTAGAAACCTCCACGTGGTGCTATTCGGGTAACGC...   \n",
       "8   GAAACACAAGATTGAAACCTCCACGTGGTGTGTTTTGGGTAAAGCT...   \n",
       "9   GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGGTAACGC...   \n",
       "10  GAATAGCAATAGTAGAAACCTCCACGTGGTGCTATTTGGGTAATGC...   \n",
       "11  GGTGCTAAAAACAAAACCTCCTCGTGGTAGCACTGGGTAATGCTAA...   \n",
       "12  GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGGTAACGC...   \n",
       "13  GATTCACATAGTAAGAACCTCCACGTGGTGTGAGTTGGGTAATGCT...   \n",
       "14  GAAGTGTGTGTAAGCAACCTCCACGCGGTACACTTCGGGTAACTCT...   \n",
       "15  GGGAACTATCCACAATACCTTCACGCGGTGTTCCTGATTTCGACTA...   \n",
       "16  GGTGCTAAAAACAAAACCTCCTCGCGGTAGCACTGGGTAATGCTAA...   \n",
       "17  GAGACACAAGATTATGTACTTCCACGCAGTGTGTCTTGGGTAATTC...   \n",
       "18  GAATAGCAAATGAGAAACCTCCACGTGGTGCTATTTGGGTAACGCT...   \n",
       "19  GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGATAACGC...   \n",
       "20  GGGAAACTATATTCTCAACCTGCTCGTGGTGTTTCCTGCTTTTACT...   \n",
       "21  GAATAGCAAATGAGAAACCTCCACGTGGTGCTATTTGGGTAATGCT...   \n",
       "22  GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGATAACGC...   \n",
       "23  GAATAGCAATAGTAGAAACCTCCACGTGGTGCTATTCGGGTAATGC...   \n",
       "24  GGTCACTAAAGTAGATACTTCCACGTAGTGTGACTGGATTAAAACT...   \n",
       "25  GACTCACAAGTAAGAACCTCCACGTGGTGTGAGTTGGATAACGCTA...   \n",
       "26  GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTCGGATAACGC...   \n",
       "27  GGTCACATAGTAGAAACCTCCACGTGGTGTGACTGGATAACCAACC...   \n",
       "28  GAATCTCATAGTCGAAACCTCCACGTGGTGAGATTCGGGTAACGCT...   \n",
       "29  GAATAGCGTAAGTAGAAACCTCCACGTGGTGCTATTTGGGTAACGC...   \n",
       "\n",
       "                             ID2  \n",
       "0    SAMEA1906581_a1_ct61757_vs1  \n",
       "1   SAMEA2580118_a1_ct141914_vs1  \n",
       "2   SAMEA2580118_a1_ct141914_vs1  \n",
       "3         SAMEA2737730_a1_ct2155  \n",
       "4       SAMN00792023_b1_ct15_vs2  \n",
       "5         SAMN00792115_a1_ct6055  \n",
       "6         SAMN05826973_a1_ct5488  \n",
       "7   SAMN06164413_a1_ct151010_vs1  \n",
       "8   SAMN06164915_a1_ct144655_vs1  \n",
       "9    SAMN09781698_a1_ct58215_vs1  \n",
       "10  SAMN09781742_a1_ct130054_vs1  \n",
       "11  SAMN09781762_a1_ct159411_vs1  \n",
       "12  SAMN09781783_a1_ct211607_vs1  \n",
       "13   SAMN10080920_a1_ct29203_vs1  \n",
       "14        SAMN11252288_a1_ct6785  \n",
       "15    SAMN11893770_a1_ct7896_vs1  \n",
       "16    SAMN11893770_a1_ct7896_vs1  \n",
       "17           SRS015484_a1_ct4036  \n",
       "18           SRS017375_a1_ct1878  \n",
       "19           SRS017375_a1_ct1878  \n",
       "20      SRS076804_a1_ct18023_vs1  \n",
       "21   SAMEA1906455_a1_ct67271_vs1  \n",
       "22   SAMEA1906455_a1_ct67271_vs1  \n",
       "23  SAMEA2580244_a1_ct185502_vs1  \n",
       "24   SAMEA2737658_a1_ct63307_vs1  \n",
       "25        SAMEA2737747_a1_ct8487  \n",
       "26  SAMEA2737752_a1_ct159360_vs1  \n",
       "27       SAMEA2737813_a1_ct27500  \n",
       "28       SAMN00791959_a1_ct12730  \n",
       "29   SAMN00993245_a1_ct62305_vs1  \n",
       "\n",
       "[30 rows x 54 columns]"
      ]
     },
     "execution_count": 41,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "merg.iloc[:30]"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
