{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Set the maximum number of columns to display to 999\n",
    "pd.options.display.max_columns = 999\n",
    "\n",
    "# Open the file containing the ribozyme locations\n",
    "rnarobo=open(\"../tables_and_result/rnarobo_result_dereplicated.txt\",\"r\")\n",
    "\n",
    "# Read the contents of the file into a list\n",
    "rnarobol=rnarobo.readlines()\n",
    "\n",
    "# Close the file\n",
    "rnarobo.close()\n",
    "\n",
    "# Create a dictionary to store the start and end positions of each ribozyme\n",
    "dicti_r={}\n",
    "\n",
    "# Loop through each line of the ribozyme file, skipping the header and footer\n",
    "for line in rnarobol[11:-6]:\n",
    "    # If the line does not start with \"|\", it contains ribozyme information\n",
    "    if line[0]!=\"|\":\n",
    "        # Replace any periods in the line with underscores\n",
    "        line=line.replace(\".\",\"_\")\n",
    "        # Split the line into a list of values\n",
    "        line=line.strip().split()\n",
    "        # If the scaffold ID is already in the dictionary, append the start and end positions to the existing list\n",
    "        if line[2] in dicti_r:\n",
    "            dicti_r[line[2]].append([int(line[0]),int(line[1])])\n",
    "        # If the scaffold ID is not in the dictionary, create a new list with the start and end positions\n",
    "        else:\n",
    "            dicti_r[line[2]]=[[int(line[0]),int(line[1])]]\n",
    "        \n",
    "# Read the coding analysis file into a pandas DataFrame\n",
    "codings=pd.read_csv(\"../alternative_coding_res/output/CD_analysis.csv\",sep=\",\")\n",
    "\n",
    "# Create new DataFrames for each predicted code\n",
    "codings_15=codings[codings[\"pred_code\"]==15]\n",
    "codings_4=codings[codings[\"pred_code\"]==4]\n",
    "codings_11=codings[codings[\"pred_code\"]==11]\n",
    "\n",
    "# Create a new empty DataFrame to store the ribozyme information\n",
    "df=pd.DataFrame(columns=[\"name\",\"start\",\"end\",\"dire\",\"code\"])\n",
    "count=0\n",
    "\n",
    "# Loop through each scaffold ID in the ribozyme dictionary\n",
    "for i in dicti_r:\n",
    "    # Get the scaffold ID\n",
    "    ids=i\n",
    "    # Loop through each start and end position for the current scaffold ID\n",
    "    for x in range(len(dicti_r[ids])):\n",
    "        # Get the start and end positions\n",
    "        start=dicti_r[ids][x][0]\n",
    "        end=dicti_r[ids][x][1]\n",
    "        # Add the scaffold ID, start position, and end position to the DataFrame\n",
    "        df.loc[count,\"name\"]=ids\n",
    "        df.iloc[count,1]=start\n",
    "        df.iloc[count,2]=end\n",
    "        # If the start position is less than the end position, set the direction to \"1\" (sense)\n",
    "        if start < end:\n",
    "            df.iloc[count,3]=\"1\"\n",
    "        # If the start position is greater than the end position, set the direction to \"-1\" (antisense)\n",
    "        else: \n",
    "            df.iloc[count,3]=\"-1\"\n",
    "         \n",
    "        # Get the predicted code for the current scaffold ID\n",
    "        try:\n",
    "            df.iloc[count,4]=codings[codings[\"Scaffold\"]==ids][\"pred_code\"].values[0]\n",
    "        # If the predicted code cannot be found, set it to \"NA\"\n",
    "        except:\n",
    "            df.iloc[count,4]=\"NA\"\n",
    "        count+=1"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "# Open the file containing the raw protein sequences for code 11\n",
    "rawfile=open(\"../alternative_coding_res/proteins.faa\")\n",
    "rl11=rawfile.readlines()\n",
    "rawfile.close()\n",
    "\n",
    "# Open the file containing the raw protein sequences for code 15\n",
    "rawfile15=open(\"../alternative_coding_res/proteins15.faa\")\n",
    "rl15=rawfile15.readlines()\n",
    "rawfile15.close()\n",
    "\n",
    "# Open the file containing the raw protein sequences for code 4\n",
    "rawfile4=open(\"../alternative_coding_res/proteins4.faa\")\n",
    "rl4=rawfile4.readlines()\n",
    "rawfile4.close()\n",
    "\n",
    "# Define a function to convert the raw protein sequences to a pandas DataFrame\n",
    "def raw_to_df_code(code):\n",
    "    # Initialize empty lists to store the values for each column in the DataFrame\n",
    "    name=[]\n",
    "    start=[]\n",
    "    end=[]\n",
    "    ids=[]\n",
    "    dire=[]\n",
    "    # Create a new empty DataFrame with the appropriate column names\n",
    "    coding_genes=pd.DataFrame(columns=[\"name\",\"ids\",\"start\",\"end\",\"dire\"])\n",
    "    # Loop through each line of the raw protein sequence file\n",
    "    for lin in code:\n",
    "        # If the line starts with \">\", it contains header information\n",
    "        if lin[0]==\">\":\n",
    "            # Get the name of the genome from the header\n",
    "            nam=lin.strip().split(\"#\")[0][1:]\n",
    "            name.append(lin.strip().split(\"#\")[0][1:])\n",
    "            # Get the ID of the coding gene from the header\n",
    "            ids.append(\"_\".join(nam.split(\"_\")[:-1]))\n",
    "            # Get the start, end, and direction of the coding gene from the header\n",
    "            start.append(lin.strip().split(\"#\")[1])\n",
    "            end.append(lin.strip().split(\"#\")[2])\n",
    "            dire.append(lin.strip().split(\"#\")[3])\n",
    "\n",
    "    # Add the values to the DataFrame\n",
    "    coding_genes[\"name\"]=name\n",
    "    coding_genes[\"ids\"]=ids\n",
    "    coding_genes[\"start\"]=start\n",
    "    coding_genes[\"end\"]=end\n",
    "    coding_genes[\"dire\"]=dire\n",
    "\n",
    "    return coding_genes\n",
    "\n",
    "# Convert the raw protein sequences to DataFrames for each code\n",
    "coding_genes_15=raw_to_df_code(rl15)\n",
    "coding_genes_11=raw_to_df_code(rl11)\n",
    "coding_genes_4=raw_to_df_code(rl4)\n",
    "\n",
    "# Create a dictionary to store the DataFrames for each code\n",
    "coding_genes_dic={}\n",
    "coding_genes_dic[\"11\"]=coding_genes_11\n",
    "coding_genes_dic[\"4\"]=coding_genes_4\n",
    "coding_genes_dic[\"15\"]=coding_genes_15\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 25,
   "metadata": {},
   "outputs": [],
   "source": [
    "#read is the theta ribozymes"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import numpy as np\n",
    "\n",
    "# Read in the combined tRNA and ribozyme sequence file as a pandas DataFrame\n",
    "theta=pd.read_csv(\"../tables_and_result/comb_trna_seq_plus_ribo_fix_march.csv\",sep=\"\\t\")\n",
    "\n",
    "# Initialize counters for each category\n",
    "cat1a=0\n",
    "cat1b=0\n",
    "cat2aa=0\n",
    "cat2ab=0\n",
    "cat2ba=0\n",
    "cat2bb=0\n",
    "cat2c=0\n",
    "cat3a=0\n",
    "cat3b=0\n",
    "cat4a=0\n",
    "cat4b=0\n",
    "thet=0\n",
    "\n",
    "# Initialize an empty list to store the smallest distances between ribozymes and genes\n",
    "small=[]\n",
    "\n",
    "# Loop through each row of the DataFrame\n",
    "for x in range(len(df)):\n",
    "    # Initialize a variable to store the smallest distance between the ribozyme and a gene\n",
    "    smallest=10000\n",
    "    # Get the code, ID, start, end, and direction of the current ribozyme\n",
    "    code=df.iloc[x,-1]\n",
    "    name=df.iloc[x,0]\n",
    "    start=int(df.iloc[x,1])\n",
    "    end=int(df.iloc[x,2])\n",
    "    dire=int(df.iloc[x,3])\n",
    "    # Select the corresponding genes from the called proteins file\n",
    "    dff=coding_genes_dic[str(code)][coding_genes_dic[str(code)][\"ids\"]==name]\n",
    "    # Initialize a variable to check if the ribozyme is a theta ribozyme\n",
    "    no=0\n",
    "    \n",
    "    # Check if the ribozyme is a theta ribozyme\n",
    "    if name in theta[\"ID\"].values:\n",
    "        tdf=theta[theta[\"ID\"]==name]\n",
    "        if start in theta[\"rnarobo_start\"].values:\n",
    "            thet+=1\n",
    "            no=2  \n",
    "            \n",
    "    # If the ribozyme is not a theta ribozyme\n",
    "    if no==0:\n",
    "        for z in range(len(dff)):\n",
    "            # Get the start, end, and direction of each called gene\n",
    "            st=int(dff.iloc[z,2])\n",
    "            en=int(dff.iloc[z,3])\n",
    "            di=int(dff.iloc[z,4])\n",
    "            # If the start of the ribozyme is within a gene\n",
    "            if start in range(st,en+1):\n",
    "                # If the end of the ribozyme is also within the gene\n",
    "                if end in range(st,en+1):\n",
    "                    # Sense\n",
    "                    if dire==di:\n",
    "                        cat1a+=1\n",
    "                        no=1\n",
    "                    # Antisense\n",
    "                    else:\n",
    "                        cat1b+=1\n",
    "                        no=1\n",
    "                # If the end of the ribozyme is not within the gene\n",
    "                else:\n",
    "                    if dire==di:\n",
    "                        cat3a+=1\n",
    "                        no=1\n",
    "                    else:\n",
    "                        cat3b+=1\n",
    "                        no=1\n",
    "            # If only the end of the ribozyme is in the gene, not the start\n",
    "            elif end in range(st,en+1):\n",
    "                # Should not be needed, but keep for consistency in code\n",
    "                if start in range(st,en+1):\n",
    "                    # Sense\n",
    "                    if dire==di:\n",
    "                        cat1a+=1\n",
    "                        no=1\n",
    "                    # Antisense\n",
    "                    else:\n",
    "                        cat1b+=1\n",
    "                        no=1\n",
    "                # Start is not in the gene\n",
    "                else:\n",
    "                    if dire==di:\n",
    "                        cat3a+=1\n",
    "                        no=1\n",
    "                    else:\n",
    "                        cat3b+=1\n",
    "                        no=1\n",
    "            # Check where the ribozyme is closest to the next gene (end or start)\n",
    "            if abs(start-st) < smallest:\n",
    "                smallest=abs(start-st)\n",
    "                if dire==di:\n",
    "                    dircheck=1\n",
    "                else:\n",
    "                    dircheck=0\n",
    "            if abs(start-en) < smallest:\n",
    "                smallest=abs(start-en)\n",
    "                if dire==di:\n",
    "                    dircheck=1\n",
    "                else:\n",
    "                    dircheck=0\n",
    "            if abs(end-st) < smallest:\n",
    "                smallest=abs(end-st)\n",
    "                if dire==di:\n",
    "                    dircheck=1\n",
    "                else:\n",
    "                    dircheck=0\n",
    "            if abs(end-en) < smallest:\n",
    "                smallest=abs(end-en)\n",
    "                if dire==di:\n",
    "                    dircheck=1\n",
    "                else:\n",
    "                    dircheck=0\n",
    "        # If the ribozyme was neither overlapping nor completely within the gene\n",
    "        if no==0:\n",
    "            # Implement length check: how far away is the closest gene (here: <=50 nucleotides away)\n",
    "            if smallest<=50:\n",
    "                # Sense\n",
    "                if dircheck==1:\n",
    "                    cat2aa+=1\n",
    "                # Antisense\n",
    "                else:\n",
    "                    cat2ab+=1\n",
    "            elif smallest<=200:\n",
    "                if dircheck==1:\n",
    "                    cat2ba+=1\n",
    "                else:\n",
    "                    cat2bb+=1\n",
    "            else:\n",
    "                if dircheck==1:\n",
    "                    cat4a+=1\n",
    "                else:\n",
    "                    cat4b+=1\n",
    "            small.append(smallest)\n",
    "\n",
    "\n",
    "#Print the occurences for each category\n",
    "\n",
    "print(\"Cat1a:\"+str(cat1a))\n",
    "print(\"Cat1b:\"+str(cat1b))\n",
    "\n",
    "print(\"Cat3a:\"+str(cat3a))\n",
    "\n",
    "print(\"Cat3b:\"+str(cat3b))\n",
    "\n",
    "print(\"Cat2aa:\"+str(cat2aa))\n",
    "\n",
    "print(\"Cat2ab:\"+str(cat2ab))\n",
    "\n",
    "print(\"Cat2ba:\"+str(cat2ba))\n",
    "print(\"Cat2bb:\"+str(cat2bb))\n",
    "\n",
    "print(\"Cat4a:\"+str(cat4a))\n",
    "print(\"Cat4b:\"+str(cat4b))\n",
    "\n",
    "print(\"Theta:\"+str(thet))\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 99,
   "metadata": {},
   "outputs": [],
   "source": [
    "#how many of the theta ribozymes are in range of +-1\n",
    "t1=theta[abs(theta[\"rnarobo_start\"]-theta[\"trna_end\"]) <2]\n",
    "t1"
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.7.6"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
